// thirdweb/dist/cjs/extensions/erc721/__generated__/Multiwrap/read/getWrappedContents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetWrappedContentsSupported = isGetWrappedContentsSupported;
exports.encodeGetWrappedContentsParams = encodeGetWrappedContentsParams;
exports.encodeGetWrappedContents = encodeGetWrappedContents;
exports.decodeGetWrappedContentsResult = decodeGetWrappedContentsResult;
exports.getWrappedContents = getWrappedContents;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xd5576d26";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_tokenId",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        name: "contents",
        components: [
            {
                type: "address",
                name: "assetContract",
            },
            {
                type: "uint8",
                name: "tokenType",
            },
            {
                type: "uint256",
                name: "tokenId",
            },
            {
                type: "uint256",
                name: "amount",
            },
        ],
    },
];
/**
 * Checks if the `getWrappedContents` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getWrappedContents` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isGetWrappedContentsSupported } from "thirdweb/extensions/erc721";
 * const supported = isGetWrappedContentsSupported(["0x..."]);
 * ```
 */
function isGetWrappedContentsSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getWrappedContents" function.
 * @param options - The options for the getWrappedContents function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeGetWrappedContentsParams } from "thirdweb/extensions/erc721";
 * const result = encodeGetWrappedContentsParams({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetWrappedContentsParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenId]);
}
/**
 * Encodes the "getWrappedContents" function into a Hex string with its parameters.
 * @param options - The options for the getWrappedContents function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeGetWrappedContents } from "thirdweb/extensions/erc721";
 * const result = encodeGetWrappedContents({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetWrappedContents(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetWrappedContentsParams(options).slice(2));
}
/**
 * Decodes the result of the getWrappedContents function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeGetWrappedContentsResult } from "thirdweb/extensions/erc721";
 * const result = decodeGetWrappedContentsResultResult("...");
 * ```
 */
function decodeGetWrappedContentsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getWrappedContents" function on the contract.
 * @param options - The options for the getWrappedContents function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { getWrappedContents } from "thirdweb/extensions/erc721";
 *
 * const result = await getWrappedContents({
 *  contract,
 *  tokenId: ...,
 * });
 *
 * ```
 */
async function getWrappedContents(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenId],
    });
}
