// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/swap/FiatValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FiatValue = FiatValue;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const cryptoToFiat_js_1 = require("../../../../../../../pay/convert/cryptoToFiat.js");
const formatNumber_js_1 = require("../../../../../../../utils/formatNumber.js");
const index_js_1 = require("../../../../../../core/design-system/index.js");
const Skeleton_js_1 = require("../../../../components/Skeleton.js");
const text_js_1 = require("../../../../components/text.js");
const useDebouncedValue_js_1 = require("../../../../hooks/useDebouncedValue.js");
const nativeToken_js_1 = require("../../nativeToken.js");
function FiatValue(props) {
    const deferredTokenAmount = (0, useDebouncedValue_js_1.useDebouncedValue)(props.tokenAmount, 500);
    const cryptoToFiatQuery = (0, react_query_1.useQuery)({
        queryKey: [
            "cryptoToFiat",
            props.chain.id,
            (0, nativeToken_js_1.getTokenAddress)(props.token),
            deferredTokenAmount,
        ],
        queryFn: () => (0, cryptoToFiat_js_1.convertCryptoToFiat)({
            client: props.client,
            chain: props.chain,
            fromTokenAddress: (0, nativeToken_js_1.getTokenAddress)(props.token),
            fromAmount: Number(deferredTokenAmount),
            to: "USD",
        }),
    });
    if (cryptoToFiatQuery.isLoading) {
        return (0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { width: "50px", height: index_js_1.fontSize.lg });
    }
    return cryptoToFiatQuery.data?.result ? ((0, jsx_runtime_1.jsxs)(text_js_1.Text, { ...props, children: ["$", (0, formatNumber_js_1.formatNumber)(cryptoToFiatQuery.data.result, 2).toFixed(2)] })) : null;
}
