// thirdweb/dist/cjs/extensions/dynamic-contracts/__generated__/IExtensionManager/write/addExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isAddExtensionSupported = isAddExtensionSupported;
exports.encodeAddExtensionParams = encodeAddExtensionParams;
exports.encodeAddExtension = encodeAddExtension;
exports.addExtension = addExtension;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xe05688fe";
const FN_INPUTS = [
    {
        type: "tuple",
        name: "extension",
        components: [
            {
                type: "tuple",
                name: "metadata",
                components: [
                    {
                        type: "string",
                        name: "name",
                    },
                    {
                        type: "string",
                        name: "metadataURI",
                    },
                    {
                        type: "address",
                        name: "implementation",
                    },
                ],
            },
            {
                type: "tuple[]",
                name: "functions",
                components: [
                    {
                        type: "bytes4",
                        name: "functionSelector",
                    },
                    {
                        type: "string",
                        name: "functionSignature",
                    },
                ],
            },
        ],
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `addExtension` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `addExtension` method is supported.
 * @extension DYNAMIC-CONTRACTS
 * @example
 * ```ts
 * import { isAddExtensionSupported } from "thirdweb/extensions/dynamic-contracts";
 *
 * const supported = isAddExtensionSupported(["0x..."]);
 * ```
 */
function isAddExtensionSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "addExtension" function.
 * @param options - The options for the addExtension function.
 * @returns The encoded ABI parameters.
 * @extension DYNAMIC-CONTRACTS
 * @example
 * ```ts
 * import { encodeAddExtensionParams } from "thirdweb/extensions/dynamic-contracts";
 * const result = encodeAddExtensionParams({
 *  extension: ...,
 * });
 * ```
 */
function encodeAddExtensionParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.extension]);
}
/**
 * Encodes the "addExtension" function into a Hex string with its parameters.
 * @param options - The options for the addExtension function.
 * @returns The encoded hexadecimal string.
 * @extension DYNAMIC-CONTRACTS
 * @example
 * ```ts
 * import { encodeAddExtension } from "thirdweb/extensions/dynamic-contracts";
 * const result = encodeAddExtension({
 *  extension: ...,
 * });
 * ```
 */
function encodeAddExtension(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeAddExtensionParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "addExtension" function on the contract.
 * @param options - The options for the "addExtension" function.
 * @returns A prepared transaction object.
 * @extension DYNAMIC-CONTRACTS
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { addExtension } from "thirdweb/extensions/dynamic-contracts";
 *
 * const transaction = addExtension({
 *  contract,
 *  extension: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
function addExtension(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.extension];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
    });
}
