// thirdweb/dist/cjs/extensions/erc1155/drops/write/claimTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.claimTo = claimTo;
exports.isClaimToSupported = isClaimToSupported;
const get_claim_params_js_1 = require("../../../../utils/extensions/drops/get-claim-params.js");
const getContractMetadata_js_1 = require("../../../common/read/getContractMetadata.js");
const GeneratedClaim = require("../../__generated__/IDrop1155/write/claim.js");
const claimCondition_js_1 = require("../../__generated__/IDropSinglePhase1155/read/claimCondition.js");
const getActiveClaimCondition_js_1 = require("../read/getActiveClaimCondition.js");
/**
 * Claim ERC1155 NFTs to a specified address
 * @param options - The options for the transaction
 * @extension ERC1155
 * @example
 * ### Basic usage
 * ```ts
 * import { claimTo } from "thirdweb/extensions/erc1155";
 * import { sendTransaction } from "thirdweb";
 *
 * const transaction = claimTo({
 *   contract,
 *   to: "0x...",
 *   tokenId: 0n,
 *   quantity: 1n,
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 *
 * ### For Drops with allowlists
 * You need to specify the claimer address as the `from` param to avoid any issue with the allowlist
 * ```ts
 * const transaction = claimTo({
 *   contract,
 *   to: "0x...",
 *   tokenId: 0n,
 *   quantity: 1n,
 *   from: "0x...", // address of the one claiming
 * });
 * ```
 * @throws If no claim condition is set
 * @returns The prepared transaction
 */
function claimTo(options) {
    return GeneratedClaim.claim({
        contract: options.contract,
        async asyncParams() {
            const params = await (0, get_claim_params_js_1.getClaimParams)({
                type: "erc1155",
                contract: options.contract,
                to: options.to,
                quantity: options.quantity,
                from: options.from,
                tokenId: options.tokenId,
                singlePhaseDrop: options.singlePhaseDrop,
            });
            return {
                ...params,
                tokenId: options.tokenId,
            };
        },
    });
}
/**
 * Checks if the `claimTo` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `claimTo` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isClaimToSupported } from "thirdweb/extensions/erc1155";
 *
 * const supported = isClaimToSupported(["0x..."]);
 * ```
 */
function isClaimToSupported(availableSelectors) {
    return [
        // has to support the claim method
        GeneratedClaim.isClaimSupported(availableSelectors),
        // requires contractMetadata for claimer proofs
        (0, getContractMetadata_js_1.isGetContractMetadataSupported)(availableSelectors),
        // required to check if the contract supports the getActiveClaimCondition method
        (0, getActiveClaimCondition_js_1.isGetActiveClaimConditionSupported)(availableSelectors) ||
            (0, claimCondition_js_1.isClaimConditionSupported)(availableSelectors),
    ].every(Boolean);
}
