// thirdweb/dist/cjs/wallets/smart/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEntryPointVersion = exports.getDefaultBundlerUrl = exports.getDefaultAccountFactory = exports.TokenPaymaster = exports.MANAGED_ACCOUNT_GAS_BUFFER = exports.ENTRYPOINT_ADDRESS_v0_7 = exports.ENTRYPOINT_ADDRESS_v0_6 = exports.DEFAULT_ACCOUNT_FACTORY_V0_7 = exports.DEFAULT_ACCOUNT_FACTORY_V0_6 = exports.DUMMY_SIGNATURE = void 0;
const address_js_1 = require("../../../utils/address.js");
const domains_js_1 = require("../../../utils/domains.js");
exports.DUMMY_SIGNATURE = "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
exports.DEFAULT_ACCOUNT_FACTORY_V0_6 = "0x85e23b94e7F5E9cC1fF78BCe78cfb15B81f0DF00";
exports.DEFAULT_ACCOUNT_FACTORY_V0_7 = "0x4be0ddfebca9a5a4a617dee4dece99e7c862dceb";
exports.ENTRYPOINT_ADDRESS_v0_6 = "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789"; // v0.6
exports.ENTRYPOINT_ADDRESS_v0_7 = "0x0000000071727De22E5E9d8BAf0edAc6f37da032"; // v0.7
exports.MANAGED_ACCOUNT_GAS_BUFFER = 50000n;
exports.TokenPaymaster = {
    BASE_USDC: {
        chainId: 8453,
        paymasterAddress: "0x2222f2738BE6bB7aA0Bfe4AEeAf2908172CF5539",
        tokenAddress: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
        balanceStorageSlot: 9n,
    },
    CELO_CUSD: {
        chainId: 42220,
        paymasterAddress: "0x3feA3c5744D715ff46e91C4e5C9a94426DfF2aF9",
        tokenAddress: "0x765DE816845861e75A25fCA122bb6898B8B1282a",
        balanceStorageSlot: 9n,
    },
    LISK_LSK: {
        chainId: 1135,
        paymasterAddress: "0x9eb8cf7fBa5ed9EeDCC97a0d52254cc0e9B1AC25",
        tokenAddress: "0xac485391EB2d7D88253a7F1eF18C37f4242D1A24",
        balanceStorageSlot: 9n,
    },
};
/*
 * @internal
 */
const getDefaultAccountFactory = (entryPointAddress) => {
    const version = (0, exports.getEntryPointVersion)(entryPointAddress || exports.ENTRYPOINT_ADDRESS_v0_6);
    if (version === "v0.7") {
        return exports.DEFAULT_ACCOUNT_FACTORY_V0_7;
    }
    return exports.DEFAULT_ACCOUNT_FACTORY_V0_6;
};
exports.getDefaultAccountFactory = getDefaultAccountFactory;
/**
 * @internal
 */
const getDefaultBundlerUrl = (chain) => {
    const domain = (0, domains_js_1.getThirdwebDomains)().bundler;
    if (domain.startsWith("localhost:")) {
        return `http://${domain}/v2?chain=${chain.id}`;
    }
    return `https://${chain.id}.${domain}/v2`;
};
exports.getDefaultBundlerUrl = getDefaultBundlerUrl;
const getEntryPointVersion = (address) => {
    const checksummedAddress = (0, address_js_1.getAddress)(address);
    if (checksummedAddress === exports.ENTRYPOINT_ADDRESS_v0_6) {
        return "v0.6";
    }
    if (checksummedAddress === exports.ENTRYPOINT_ADDRESS_v0_7) {
        return "v0.7";
    }
    throw new Error("Unknown paymaster version");
};
exports.getEntryPointVersion = getEntryPointVersion;
