// thirdweb/dist/esm/react/native/ui/prebuilt/Account/avatar.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import { Image } from "react-native";
import { resolveAvatar } from "../../../../../extensions/ens/resolve-avatar.js";
import { resolveName, } from "../../../../../extensions/ens/resolve-name.js";
import { getSocialProfiles } from "../../../../../social/profiles.js";
import { parseAvatarRecord } from "../../../../../utils/ens/avatar.js";
import { useAccountContext } from "../../../../core/account/provider.js";
/**
 * The component for showing the avatar of the account.
 * If fetches all the social profiles linked to your wallet, including: Farcaster, ENS, Lens (more to be added)
 * You can choose which social profile you want to display. Defaults to the first item in the list.
 *
 * @example
 * ### Basic usage
 * ```tsx
 * import { AccountProvider, AccountAvatar } from "thirdweb/react";
 *
 * <AccountProvider address="0x...">
 *   <AccountAvatar />
 * </AccountProvider>
 * ```
 * Result: An <img /> component, if the avatar is resolved successfully
 * ```html
 * <img alt="" src="resolved-url-for-the-avatar" />
 * ```
 *
 * ### Show a loading sign when the avatar is being resolved
 * ```tsx
 * import { AccountProvider, AccountAvatar } from "thirdweb/react";
 *
 * <AccountProvider address="0x...">
 *   <AccountAvatar
 *     loadingComponent={<YourLoadingComponent />}
 *   />
 * </AccountProvider>
 * ```
 *
 * ### Fallback to something when the avatar fails to resolve
 * ```tsx
 * import { AccountProvider, AccountAvatar } from "thirdweb/react";
 *
 * <AccountProvider address="0x...">
 *   <AccountAvatar
 *     fallbackComponent={<DummyImage />}
 *   />
 * </AccountProvider>
 * ```
 *
 * ### Select a social profile to display
 * If you wallet associates with more than one social profiles (Lens, Farcaster, ENS, etc.)
 * You can specify which service you want to prioritize using the `socialType` props
 * ```tsx
 * import { AccountProvider, AccountAvatar } from "thirdweb/react";
 *
 * <AccountProvider address="0x...">
 *   <AccountAvatar
 *     // Choose between: "farcaster" | "lens" | "ens"
 *     socialType={"ens"}
 *   />
 * </AccountProvider>
 * ```
 *
 * ### Custom ENS resolver chain
 * This component shares the same props with the ENS extension `resolveAvatar`
 * ```tsx
 * import { AccountProvider, AccountAvatar } from "thirdweb/react";
 * import { base } from "thirdweb/chains";
 *
 * <AccountProvider address="0x...">
 *   <AccountAvatar
 *     resolverAddress={"0x..."}
 *     resolverChain={base}
 *   />
 * </AccountProvider>
 * ```
 *
 * ### Custom query options for useQuery
 * This component uses `@tanstack-query`'s useQuery internally.
 * You can use the `queryOptions` prop for more fine-grained control
 * ```tsx
 * <AccountAvatar
 *   queryOptions={{
 *     enabled: isEnabled,
 *     retry: 3,
 *   }}
 * />
 * ```
 * @returns An <img /> if the avatar is resolved successfully
 * @component
 * @wallet
 * @beta
 */
export function AccountAvatar({ socialType, resolverAddress, resolverChain, loadingComponent, fallbackComponent, queryOptions, ...restProps }) {
    const { address, client } = useAccountContext();
    const avatarQuery = useQuery({
        queryKey: [
            "account-avatar",
            address,
            { socialType },
            { resolverAddress, resolverChain },
        ],
        queryFn: async () => {
            const [socialData, ensName] = await Promise.all([
                getSocialProfiles({ address, client }),
                resolveName({
                    client,
                    address: address || "",
                    resolverAddress,
                    resolverChain,
                }),
            ]);
            const uri = socialData?.filter((p) => p.avatar && (socialType ? p.type === socialType : true))[0]?.avatar;
            const [resolvedSocialAvatar, resolvedENSAvatar] = await Promise.all([
                uri ? parseAvatarRecord({ client, uri }) : undefined,
                ensName
                    ? resolveAvatar({
                        client,
                        name: ensName,
                    })
                    : undefined,
            ]);
            // If no social image + ens name found -> exit and show <Blobbie />
            if (!resolvedSocialAvatar && !resolvedENSAvatar) {
                throw new Error("Failed to resolve social + ens avatar");
            }
            // else, prioritize the social image first
            if (resolvedSocialAvatar) {
                return resolvedSocialAvatar;
            }
            if (resolvedENSAvatar) {
                return resolvedENSAvatar;
            }
            throw new Error("Failed to resolve social + ens avatar");
        },
        retry: false,
        ...queryOptions,
    });
    if (avatarQuery.isLoading) {
        return loadingComponent || null;
    }
    if (!avatarQuery.data) {
        return fallbackComponent || null;
    }
    return _jsx(Image, { source: { uri: avatarQuery.data }, ...restProps });
}
