// thirdweb/dist/cjs/react/web/ui/SiteEmbed.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SiteEmbed = SiteEmbed;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const storage_js_1 = require("../../../react/core/utils/storage.js");
const webStorage_js_1 = require("../../../utils/storage/webStorage.js");
const is_ecosystem_wallet_js_1 = require("../../../wallets/ecosystem/is-ecosystem-wallet.js");
const client_scoped_storage_js_1 = require("../../../wallets/in-app/core/authentication/client-scoped-storage.js");
const useActiveWallet_js_1 = require("../../core/hooks/wallets/useActiveWallet.js");
/**
 * Embeds another thirdweb-supported site for seamless in-app and ecosystem wallet connection.
 *
 *  Make sure the embedded site includes <AutoConnect /> and supports frame ancestors, see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors) for more information.
 *
 *  The embedded site must support the connected wallet (ecosystem or in-app).
 *
 * @param {Object} props - The props to pass to the iframe
 * @param {String} props.src - The URL of the site to embed
 * @param {ThirdwebClient} props.client - The current site's thirdweb client
 * @param {Ecosystem} [props.ecosystem] - The ecosystem to use for the wallet connection in the embedded site
 *
 * @example
 * ```tsx
 * import { SiteEmbed } from "thirdweb/react";
 *
 * <SiteEmbed src="https://thirdweb.com" client={thirdwebClient} ecosystem={{ id: "ecosystem.thirdweb" }} />
 * ```
 * @walletConnection
 */
function SiteEmbed({ src, client, ecosystem, ...props }) {
    if (!client.clientId) {
        throw new Error("The SiteEmbed client must have a clientId");
    }
    const activeWallet = (0, useActiveWallet_js_1.useActiveWallet)();
    const walletId = activeWallet?.id;
    const { data: { authProvider, authCookie } = {}, } = (0, react_query_1.useQuery)({
        queryKey: ["site-embed", walletId, src, client.clientId, ecosystem],
        enabled: activeWallet &&
            ((0, is_ecosystem_wallet_js_1.isEcosystemWallet)(activeWallet) ||
                walletId === "inApp" ||
                walletId === "smart"),
        queryFn: async () => {
            const storage = new client_scoped_storage_js_1.ClientScopedStorage({
                storage: webStorage_js_1.webLocalStorage,
                clientId: client.clientId,
                ecosystem,
            });
            const authProvider = await (0, storage_js_1.getLastAuthProvider)(webStorage_js_1.webLocalStorage);
            const authCookie = await storage.getAuthCookie();
            return { authProvider, authCookie };
        },
    });
    const url = new URL(src);
    if (walletId) {
        url.searchParams.set("walletId", walletId === "smart" ? "inApp" : walletId);
    }
    if (authProvider) {
        url.searchParams.set("authProvider", authProvider);
    }
    if (authCookie) {
        url.searchParams.set("authCookie", authCookie);
    }
    return ((0, jsx_runtime_1.jsx)("iframe", { src: encodeURI(url.toString()), width: "100%", height: "100%", allowFullScreen: true, ...props }));
}
