// @hey-api/openapi-ts/dist/index.js
import {createRequire}from'module';import {b,a,c,m,x,t,j,o,f as f$1,g,r,q,s,e,p as p$1,k,d,l,h,n,i}from'./chunk-R52Q6I2F.js';export{u as clientDefaultConfig,v as clientPluginHandler,w as defaultPlugins}from'./chunk-R52Q6I2F.js';import y from'node:fs';import R from'node:path';import $ from'typescript';import {fileURLToPath}from'node:url';import p from'handlebars';createRequire(import.meta.url);
var we=b((jr,Oe)=>{Oe.exports=Te;Te.sync=Rt;var qe=a("fs");function yt(e,n){var s=n.pathExt!==void 0?n.pathExt:process.env.PATHEXT;if(!s||(s=s.split(";"),s.indexOf("")!==-1))return  true;for(var r=0;r<s.length;r++){var t=s[r].toLowerCase();if(t&&e.substr(-t.length).toLowerCase()===t)return  true}return  false}function Ae(e,n,s){return !e.isSymbolicLink()&&!e.isFile()?false:yt(n,s)}function Te(e,n,s){qe.stat(e,function(r,t){s(r,r?false:Ae(t,e,n));});}function Rt(e,n){return Ae(qe.statSync(e),e,n)}});var De=b((Nr,He)=>{He.exports=Ee;Ee.sync=vt;var ke=a("fs");function Ee(e,n,s){ke.stat(e,function(r,t){s(r,r?false:Se(t,n));});}function vt(e,n){return Se(ke.statSync(e),n)}function Se(e,n){return e.isFile()&&Pt(e,n)}function Pt(e,n){var s=e.mode,r=e.uid,t=e.gid,o=n.uid!==void 0?n.uid:process.getuid&&process.getuid(),a=n.gid!==void 0?n.gid:process.getgid&&process.getgid(),i=parseInt("100",8),u=parseInt("010",8),l=parseInt("001",8),c=i|u,m=s&l||s&u&&t===a||s&i&&r===o||s&c&&o===0;return m}});var Be=b((Fr,Ie)=>{a("fs");var N;process.platform==="win32"||global.TESTING_WINDOWS?N=we():N=De();Ie.exports=V;V.sync=xt;function V(e,n,s){if(typeof n=="function"&&(s=n,n={}),!s){if(typeof Promise!="function")throw new TypeError("callback not provided");return new Promise(function(r,t){V(e,n||{},function(o,a){o?t(o):r(a);});})}N(e,n||{},function(r,t){r&&(r.code==="EACCES"||n&&n.ignoreErrors)&&(r=null,t=false),s(r,t);});}function xt(e,n){try{return N.sync(e,n||{})}catch(s){if(n&&n.ignoreErrors||s.code==="EACCES")return  false;throw s}}});var We=b((Lr,Le)=>{var w=process.platform==="win32"||process.env.OSTYPE==="cygwin"||process.env.OSTYPE==="msys",$e=a("path"),bt=w?";":":",je=Be(),Ne=e=>Object.assign(new Error(`not found: ${e}`),{code:"ENOENT"}),_e=(e,n)=>{let s=n.colon||bt,r=e.match(/\//)||w&&e.match(/\\/)?[""]:[...w?[process.cwd()]:[],...(n.path||process.env.PATH||"").split(s)],t=w?n.pathExt||process.env.PATHEXT||".EXE;.CMD;.BAT;.COM":"",o=w?t.split(s):[""];return w&&e.indexOf(".")!==-1&&o[0]!==""&&o.unshift(""),{pathEnv:r,pathExt:o,pathExtExe:t}},Fe=(e,n,s)=>{typeof n=="function"&&(s=n,n={}),n||(n={});let{pathEnv:r,pathExt:t,pathExtExe:o}=_e(e,n),a=[],i=l=>new Promise((c,m)=>{if(l===r.length)return n.all&&a.length?c(a):m(Ne(e));let d=r[l],q=/^".*"$/.test(d)?d.slice(1,-1):d,b=$e.join(q,e),O=!q&&/^\.[\\\/]/.test(e)?e.slice(0,2)+b:b;c(u(O,l,0));}),u=(l,c,m)=>new Promise((d,q)=>{if(m===t.length)return d(i(c+1));let b=t[m];je(l+b,{pathExt:o},(O,E)=>{if(!O&&E)if(n.all)a.push(l+b);else return d(l+b);return d(u(l,c,m+1))});});return s?i(0).then(l=>s(null,l),s):i(0)},Ct=(e,n)=>{n=n||{};let{pathEnv:s,pathExt:r,pathExtExe:t}=_e(e,n),o=[];for(let a=0;a<s.length;a++){let i=s[a],u=/^".*"$/.test(i)?i.slice(1,-1):i,l=$e.join(u,e),c=!u&&/^\.[\\\/]/.test(e)?e.slice(0,2)+l:l;for(let m=0;m<r.length;m++){let d=c+r[m];try{if(je.sync(d,{pathExt:t}))if(n.all)o.push(d);else return d}catch{}}}if(n.all&&o.length)return o;if(n.nothrow)return null;throw Ne(e)};Le.exports=Fe;Fe.sync=Ct;});var Ue=b((Wr,G)=>{var Me=(e={})=>{let n=e.env||process.env;return (e.platform||process.platform)!=="win32"?"PATH":Object.keys(n).reverse().find(r=>r.toUpperCase()==="PATH")||"Path"};G.exports=Me;G.exports.default=Me;});var Ke=b((Mr,ze)=>{var Ve=a("path"),qt=We(),At=Ue();function Ge(e,n){let s=e.options.env||process.env,r=process.cwd(),t=e.options.cwd!=null,o=t&&process.chdir!==void 0&&!process.chdir.disabled;if(o)try{process.chdir(e.options.cwd);}catch{}let a;try{a=qt.sync(e.command,{path:s[At({env:s})],pathExt:n?Ve.delimiter:void 0});}catch{}finally{o&&process.chdir(r);}return a&&(a=Ve.resolve(t?e.options.cwd:"",a)),a}function Tt(e){return Ge(e)||Ge(e,true)}ze.exports=Tt;});var Je=b((Ur,K)=>{var z=/([()\][%!^"`<>&|;, *?])/g;function Ot(e){return e=e.replace(z,"^$1"),e}function wt(e,n){return e=`${e}`,e=e.replace(/(?=(\\+?)?)\1"/g,'$1$1\\"'),e=e.replace(/(?=(\\+?)?)\1$/,"$1$1"),e=`"${e}"`,e=e.replace(z,"^$1"),n&&(e=e.replace(z,"^$1")),e}K.exports.command=Ot;K.exports.argument=wt;});var Xe=b((Vr,Qe)=>{Qe.exports=/^#!(.*)/;});var Ze=b((Gr,Ye)=>{var kt=Xe();Ye.exports=(e="")=>{let n=e.match(kt);if(!n)return null;let[s,r]=n[0].replace(/#! ?/,"").split(" "),t=s.split("/").pop();return t==="env"?r:r?`${t} ${r}`:t};});var nn=b((zr,en)=>{var J=a("fs"),Et=Ze();function St(e){let s=Buffer.alloc(150),r;try{r=J.openSync(e,"r"),J.readSync(r,s,0,150,0),J.closeSync(r);}catch{}return Et(s.toString())}en.exports=St;});var sn=b((Kr,on)=>{var Ht=a("path"),tn=Ke(),rn=Je(),Dt=nn(),It=process.platform==="win32",Bt=/\.(?:com|exe)$/i,$t=/node_modules[\\/].bin[\\/][^\\/]+\.cmd$/i;function jt(e){e.file=tn(e);let n=e.file&&Dt(e.file);return n?(e.args.unshift(e.file),e.command=n,tn(e)):e.file}function Nt(e){if(!It)return e;let n=jt(e),s=!Bt.test(n);if(e.options.forceShell||s){let r=$t.test(n);e.command=Ht.normalize(e.command),e.command=rn.command(e.command),e.args=e.args.map(o=>rn.argument(o,r));let t=[e.command].concat(e.args).join(" ");e.args=["/d","/s","/c",`"${t}"`],e.command=process.env.comspec||"cmd.exe",e.options.windowsVerbatimArguments=true;}return e}function _t(e,n,s){n&&!Array.isArray(n)&&(s=n,n=null),n=n?n.slice(0):[],s=Object.assign({},s);let r={command:e,args:n,options:s,file:void 0,original:{command:e,args:n}};return s.shell?r:Nt(r)}on.exports=_t;});var un=b((Jr,ln)=>{var Q=process.platform==="win32";function X(e,n){return Object.assign(new Error(`${n} ${e.command} ENOENT`),{code:"ENOENT",errno:"ENOENT",syscall:`${n} ${e.command}`,path:e.command,spawnargs:e.args})}function Ft(e,n){if(!Q)return;let s=e.emit;e.emit=function(r,t){if(r==="exit"){let o=an(t,n);if(o)return s.call(e,"error",o)}return s.apply(e,arguments)};}function an(e,n){return Q&&e===1&&!n.file?X(n.original,"spawn"):null}function Lt(e,n){return Q&&e===1&&!n.file?X(n.original,"spawnSync"):null}ln.exports={hookChildProcess:Ft,verifyENOENT:an,verifyENOENTSync:Lt,notFoundError:X};});var mn=b((Qr,k)=>{var pn=a("child_process"),Y=sn(),Z=un();function cn(e,n,s){let r=Y(e,n,s),t=pn.spawn(r.command,r.args,r.options);return Z.hookChildProcess(t,r),t}function Wt(e,n,s){let r=Y(e,n,s),t=pn.spawnSync(r.command,r.args,r.options);return t.error=t.error||Z.verifyENOENTSync(t.status,r),t}k.exports=cn;k.exports.spawn=cn;k.exports.sync=Wt;k.exports._parse=Y;k.exports._enoent=Z;});var he=async({context:e})=>{await e.broadcast("before");for(let n of e.ir.servers??[])await e.broadcast("server",{server:n});if(e.ir.components){for(let n in e.ir.components.schemas){let s=e.ir.components.schemas[n],r=`#/components/schemas/${n}`;await e.broadcast("schema",{$ref:r,name:n,schema:s});}for(let n in e.ir.components.parameters){let s=e.ir.components.parameters[n],r=`#/components/parameters/${n}`;await e.broadcast("parameter",{$ref:r,name:n,parameter:s});}for(let n in e.ir.components.requestBodies){let s=e.ir.components.requestBodies[n],r=`#/components/requestBodies/${n}`;await e.broadcast("requestBody",{$ref:r,name:n,requestBody:s});}}for(let n in e.ir.paths){let s=e.ir.paths[n];for(let r in s){let t=r,o=s[t];await e.broadcast("operation",{method:t,operation:o,path:n});}}await e.broadcast("after");};var B=e=>{switch(e){case "legacy/angular":return "AngularHttpRequest";case "legacy/axios":return "AxiosHttpRequest";case "legacy/fetch":return "FetchHttpRequest";case "legacy/node":return "NodeHttpRequest";case "legacy/xhr":return "XHRHttpRequest";default:return ""}};var ye=async(e$1,n,s,r)=>{let t=e(),o=d(t),a=r.client({$config:t,...s,httpRequest:B(o.name),models:i(s.models),services:i(s.services)});g(t)&&(j(n),y.writeFileSync(R.resolve(n,`${g(t)}.ts`),a));};var Re=async(e$1,n,s)=>{let r=e();if(r.exportCore){let t=d(r),o={httpRequest:B(t.name),server:r.base!==void 0?r.base:n.server,version:n.version};if(y.rmSync(R.resolve(e$1),{force:true,recursive:true}),y.mkdirSync(R.resolve(e$1),{recursive:true}),await y.writeFileSync(R.resolve(e$1,"OpenAPI.ts"),s.core.settings({$config:r,...o})),await y.writeFileSync(R.resolve(e$1,"ApiError.ts"),s.core.apiError({$config:r,...o})),await y.writeFileSync(R.resolve(e$1,"ApiRequestOptions.ts"),s.core.apiRequestOptions({$config:r,...o})),await y.writeFileSync(R.resolve(e$1,"ApiResult.ts"),s.core.apiResult({$config:r,...o})),t.name!=="legacy/angular"&&await y.writeFileSync(R.resolve(e$1,"CancelablePromise.ts"),s.core.cancelablePromise({$config:r,...o})),await y.writeFileSync(R.resolve(e$1,"request.ts"),s.core.request({$config:r,...o})),g(r)&&(await y.writeFileSync(R.resolve(e$1,"BaseHttpRequest.ts"),s.core.baseHttpRequest({$config:r,...o})),await y.writeFileSync(R.resolve(e$1,`${o.httpRequest}.ts`),s.core.httpRequest({$config:r,...o}))),r.request){let a=R.resolve(process.cwd(),r.request);if(!await y.existsSync(a))throw new Error(`Custom request file "${a}" does not exists`);await y.copyFileSync(a,R.resolve(e$1,"request.ts"));}}};var ve=({files:e$1})=>{let n$1=e();e$1.index=new n({dir:n$1.output.path,id:"index",name:"index.ts"}),g(n$1)&&e$1.index.add(h.exportNamedDeclaration({exports:g(n$1),module:`./${g(n$1)}`})),n$1.exportCore&&(e$1.index.add(h.exportNamedDeclaration({exports:"ApiError",module:"./core/ApiError"})),n$1.plugins["@hey-api/sdk"]?.response==="response"&&e$1.index.add(h.exportNamedDeclaration({exports:{asType:true,name:"ApiResult"},module:"./core/ApiResult"})),g(n$1)&&e$1.index.add(h.exportNamedDeclaration({exports:"BaseHttpRequest",module:"./core/BaseHttpRequest"})),d(n$1).name!=="legacy/angular"&&e$1.index.add(h.exportNamedDeclaration({exports:["CancelablePromise","CancelError"],module:"./core/CancelablePromise"})),e$1.index.add(h.exportNamedDeclaration({exports:["OpenAPI",{asType:true,name:"OpenAPIConfig"}],module:"./core/OpenAPI"}))),Object.keys(e$1).sort().forEach(s=>{let r=e$1[s];s==="index"||r.isEmpty()||["sdk","types"].includes(s)&&e$1.index.add(h.exportAllDeclaration({module:`./${r.nameWithoutExtension()}`}));});};var gt=fileURLToPath(import.meta.url),xe=R.dirname(gt),M=e=>{if(e==="off")return null;if(e){let s=R.isAbsolute(e)?e:R.resolve(xe,e);return y.existsSync(s)?s:null}let n=xe;for(;n!==R.dirname(n);){let r=y.readdirSync(n).filter(t=>t.startsWith("tsconfig")&&t.endsWith(".json")).sort(t=>t==="tsconfig.json"?-1:1);if(r[0])return R.join(n,r[0]);n=R.dirname(n);}return null},U=e=>{if(!e)return null;let n=$.readConfigFile(e,$.sys.readFile);if(n.error)throw new Error(`Couldn't read tsconfig from path: ${e}`);return $.parseJsonConfigFileContent(n.config,$.sys,R.dirname(e))};var be=async({client:e$1,openApi:n$1,templates:s})=>{let r=e(),t=n$1;if(e$1){if(r.plugins["@hey-api/sdk"]?.include&&r.plugins["@hey-api/sdk"].asClass){let l=new RegExp(r.plugins["@hey-api/sdk"].include);e$1.services=e$1.services.filter(c=>l.test(c.name));}if(r.plugins["@hey-api/typescript"]?.include){let l=new RegExp(r.plugins["@hey-api/typescript"].include);e$1.models=e$1.models.filter(c=>l.test(c.name));}}let o=R.resolve(r.output.path);r.output.clean&&k(o);let a=d(r);!f$1(r)&&"bundle"in a&&a.bundle&&l({outputPath:o,plugin:a}),await ye(t,o,e$1,s),await Re(R.resolve(r.output.path,"core"),e$1,s);let i={};for(let l of r.pluginOrder){let c=r.plugins[l],m=(c.output??"").split("/"),d=R.resolve(r.output.path,...m.slice(0,m.length-1));i[c.name]=new n({dir:d,id:`legacy-unused-${c.name}`,name:`${m[m.length-1]}.ts`}),c._handlerLegacy({client:e$1,files:i,openApi:t,plugin:c});}ve({files:i});let u=U(M(r.output.tsConfigPath));Object.entries(i).forEach(([l,c])=>{r.dryRun||(l==="index"?c.write(`
`,u):c.write(`

`,u));});},Ce=async({context:e})=>{let n=R.resolve(e.config.output.path);e.config.output.clean&&k(n);let s=d(e.config);"bundle"in s&&s.bundle&&l({outputPath:n,plugin:s});for(let r of e.config.pluginOrder){let t=e.config.plugins[r];t._handler({context:e,plugin:t});}if(await he({context:e}),!e.config.dryRun){let r=e.createFile({id:"_index",path:"index"}),t=U(M(e.config.output.tsConfigPath)),o=t?.options.moduleResolution===$.ModuleResolutionKind.NodeNext;for(let a of Object.values(e.files))if(a.nameWithoutExtension()!==r.nameWithoutExtension()){if(!a.isEmpty()&&a.exportFromIndex&&e.config.output.indexFile){let u=r.relativePathToFile({context:e,id:a.id});o&&(u.startsWith("./")||u.startsWith("../"))&&(u=`${u}.js`),r.add(h.exportAllDeclaration({module:u}));}a.write(`

`,t);}e.config.output.indexFile&&r.write(`
`,t);}};var ee=c(mn(),1),Mt={biome:{args:e=>["format","--write",e],command:"biome",name:"Biome (Format)"},prettier:{args:e=>["--ignore-unknown",e,"--write","--ignore-path","./.prettierignore"],command:"prettier",name:"Prettier"}},Ut={biome:{args:e=>["lint","--apply",e],command:"biome",name:"Biome (Lint)"},eslint:{args:e=>[e,"--fix"],command:"eslint",name:"ESLint"},oxlint:{args:e=>["--fix",e],command:"oxlint",name:"oxlint"}},fn=({config:e})=>{if(e.output.format){let n=Mt[e.output.format];console.log(`\u2728 Running ${n.name}`),(0, ee.sync)(n.command,n.args(e.output.path));}if(e.output.lint){let n=Ut[e.output.lint];console.log(`\u2728 Running ${n.name}`),(0, ee.sync)(n.command,n.args(e.output.path));}};var dn=e=>`${e}-end`,ne=e=>`${e}-length`,gn=e=>`${e}-start`,f={clear:()=>{performance.clearMarks(),performance.clearMeasures();},end:e=>performance.mark(dn(e)),getEntriesByName:e=>performance.getEntriesByName(ne(e)),measure:e=>performance.measure(ne(e),gn(e),dn(e)),start:e=>performance.mark(gn(e))},_=class{totalMeasure;constructor({totalMark:n}){this.totalMeasure=f.measure(n);}report({marks:n}){let s=Math.ceil(this.totalMeasure.duration*100)/100,r=this.totalMeasure.name;console.warn(`${r.substring(0,r.length-ne("").length)}: ${s.toFixed(2)}ms`),n.forEach(t=>{try{let o=f.measure(t),a=Math.ceil(o.duration*100)/100,i=Math.ceil(o.duration/this.totalMeasure.duration*100*100)/100;console.warn(`${t}: ${a.toFixed(2)}ms (${i.toFixed(2)}%)`);}catch{}});}};var te=e=>e.startsWith("https://get.heyapi.dev"),Gt=e=>{let n={path:""};if(e.path&&(typeof e.path!="string"||!te(e.path)))return n.path=e.path,n;let[s,r]=e.path.split("?"),o=(r||"").split("&").map(v=>v.split("=")),a=s||"";a.endsWith("/")&&(a=a.slice(0,a.length-1));let[,i]=a.split("://"),[u,l,c]=(i||"").split("/");n.organization=l||e.organization,n.project=c||e.project;let m=[],d="api_key";n.api_key=o.find(([v])=>v===d)?.[1]||e.api_key||process.env.HEY_API_TOKEN,n.api_key&&m.push(`${d}=${n.api_key}`);let q="branch";n.branch=o.find(([v])=>v===q)?.[1]||e.branch,n.branch&&m.push(`${q}=${n.branch}`);let b="commit_sha";n.commit_sha=o.find(([v])=>v===b)?.[1]||e.commit_sha,n.commit_sha&&m.push(`${b}=${n.commit_sha}`);let O="tags";n.tags=o.find(([v])=>v===O)?.[1]?.split(",")||e.tags,n.tags?.length&&m.push(`${O}=${n.tags.join(",")}`);let E="version";if(n.version=o.find(([v])=>v===E)?.[1]||e.version,n.version&&m.push(`${E}=${n.version}`),!n.organization)throw new Error("\u{1F6AB} missing organization - from which Hey API platform organization do you want to generate your output?");if(!n.project)throw new Error("\u{1F6AB} missing project - from which Hey API platform project do you want to generate your output?");let oe=m.join("&"),se=u||"get.heyapi.dev",ie=se.startsWith("localhost"),ae=[ie?"http":"https",se].join("://"),le=ie?[ae,"v1","get",n.organization,n.project].join("/"):[ae,n.organization,n.project].join("/");return n.path=oe?`${le}?${oe}`:le,n},zt=({config:e,inputPath:n,watch:s})=>{if(e.logs.level==="silent")return;s&&console.clear();let r=s?"Input changed, generating from":"Generating from";if(typeof n.path=="string"){let t=te(n.path)?`${n.organization}/${n.project}`:n.path;console.log(`\u23F3 ${r} ${t}`),te(n.path)&&(n.branch&&console.log(`branch: ${n.branch}`),n.commit_sha&&console.log(`commit: ${n.commit_sha}`),n.tags?.length&&console.log(`tags: ${n.tags.join(", ")}`),n.version&&console.log(`version: ${n.version}`));}else console.log(`\u23F3 ${r} raw OpenAPI specification`);},re=async({config:e,templates:n,watch:s$1})=>{let r$1=Gt(e.input),t=e.watch.timeout,o$1=s$1||{headers:new Headers};zt({config:e,inputPath:r$1,watch:!!s$1}),f.start("spec");let{data:a,error:i,response:u}=await o({fetchOptions:e.input.fetch,inputPath:r$1.path,timeout:t,watch:o$1});if(f.end("spec"),i&&!s$1)throw new Error(`Request failed with status ${u.status}: ${u.statusText}`);let l,c;if(a){if(f.start("parser"),e.experimentalParser&&!f$1(e)&&!g(e)&&(c=r({config:e,spec:a})),!c){let m=q({openApi:a});l=s(m,e);}if(f.end("parser"),f.start("generator"),c?await Ce({context:c}):l&&await be({client:l,openApi:a,templates:n}),f.end("generator"),f.start("postprocess"),!e.dryRun&&(fn({config:e}),e.logs.level!=="silent")){let m=process.env.INIT_CWD?`./${R.relative(process.env.INIT_CWD,e.output.path)}`:e.output.path;console.log(`\u{1F680} Done! Your output is in ${m}`);}f.end("postprocess");}return e.watch.enabled&&typeof r$1.path=="string"&&setTimeout(()=>{re({config:e,templates:n,watch:o$1});},e.watch.interval),c||l};var hn={1:function(e,n,s,r,t){return `import { NgModule} from '@angular/core';
import { HttpClientModule } from '@angular/common/http';

import { AngularHttpRequest } from './core/AngularHttpRequest';
import { BaseHttpRequest } from './core/BaseHttpRequest';
import type { OpenAPIConfig } from './core/OpenAPI';
import { OpenAPI } from './core/OpenAPI';
import { Interceptors } from './core/OpenAPI';
`},3:function(e,n,s,r,t){var o,a=e.strict,i=e.lambda;return `import type { BaseHttpRequest } from './core/BaseHttpRequest';
import type { OpenAPIConfig } from './core/OpenAPI';
import { Interceptors } from './core/OpenAPI';
import { `+((o=i(a(n,"httpRequest",{start:{line:14,column:12},end:{line:14,column:23}}),n))!=null?o:"")+" } from './core/"+((o=i(a(n,"httpRequest",{start:{line:14,column:45},end:{line:14,column:56}}),n))!=null?o:"")+`';
`},5:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return (o=a(s,"each").call(n??(e.nullContext||{}),a(n,"services"),{name:"each",hash:{},fn:e.program(6,t,0),inverse:e.noop,data:t,loc:{start:{line:18,column:0},end:{line:20,column:9}}}))!=null?o:""},6:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return "import { "+((o=a(s,"transformServiceName").call(n??(e.nullContext||{}),a(n,"name"),{name:"transformServiceName",hash:{},data:t,loc:{start:{line:19,column:9},end:{line:19,column:40}}}))!=null?o:"")+` } from './sdk.gen';
`},8:function(e,n,s,r,t){var o,a=e.strict,i=e.lambda,u=e.lookupProperty||function(l,c){if(Object.prototype.hasOwnProperty.call(l,c))return l[c]};return `@NgModule({
	imports: [HttpClientModule],
	providers: [
		{
			provide: OpenAPI,
			useValue: {
				BASE: OpenAPI?.BASE ?? '`+((o=i(a(n,"server",{start:{line:30,column:31},end:{line:30,column:37}}),n))!=null?o:"")+`',
				VERSION: OpenAPI?.VERSION ?? '`+((o=i(a(n,"version",{start:{line:31,column:37},end:{line:31,column:44}}),n))!=null?o:"")+`',
				WITH_CREDENTIALS: OpenAPI?.WITH_CREDENTIALS ?? false,
				CREDENTIALS: OpenAPI?.CREDENTIALS ?? 'include',
				TOKEN: OpenAPI?.TOKEN,
				USERNAME: OpenAPI?.USERNAME,
				PASSWORD: OpenAPI?.PASSWORD,
				HEADERS: OpenAPI?.HEADERS,
				ENCODE_PATH: OpenAPI?.ENCODE_PATH,
				interceptors: {
					response: OpenAPI?.interceptors?.response ?? new Interceptors(),
				},
			} as OpenAPIConfig,
		},
		{
			provide: BaseHttpRequest,
			useClass: AngularHttpRequest,
		},
`+((o=u(s,"each").call(n??(e.nullContext||{}),u(n,"services"),{name:"each",hash:{},fn:e.program(9,t,0),inverse:e.noop,data:t,loc:{start:{line:48,column:2},end:{line:50,column:11}}}))!=null?o:"")+`	]
})
export class `+((o=i(a(u(u(t,"root"),"$config"),"name",{start:{line:53,column:16},end:{line:53,column:34}}),n))!=null?o:"")+` {}
`},9:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return "		"+((o=a(s,"transformServiceName").call(n??(e.nullContext||{}),a(n,"name"),{name:"transformServiceName",hash:{},data:t,loc:{start:{line:49,column:2},end:{line:49,column:33}}}))!=null?o:"")+`,
`},11:function(e,n,s,r,t){var o,a=e.strict,i=e.lambda,u=n??(e.nullContext||{}),l=e.lookupProperty||function(c,m){if(Object.prototype.hasOwnProperty.call(c,m))return c[m]};return `type HttpRequestConstructor = new (config: OpenAPIConfig) => BaseHttpRequest;

export class `+((o=i(a(l(l(t,"root"),"$config"),"name",{start:{line:57,column:16},end:{line:57,column:34}}),n))!=null?o:"")+` {

`+((o=l(s,"each").call(u,l(n,"services"),{name:"each",hash:{},fn:e.program(12,t,0),inverse:e.noop,data:t,loc:{start:{line:59,column:1},end:{line:61,column:10}}}))!=null?o:"")+`
	public readonly request: BaseHttpRequest;

	constructor(config?: Partial<OpenAPIConfig>, HttpRequest: HttpRequestConstructor = `+((o=i(a(n,"httpRequest",{start:{line:65,column:87},end:{line:65,column:98}}),n))!=null?o:"")+`) {
		this.request = new HttpRequest({
			BASE: config?.BASE ?? '`+((o=i(a(n,"server",{start:{line:67,column:29},end:{line:67,column:35}}),n))!=null?o:"")+`',
			VERSION: config?.VERSION ?? '`+((o=i(a(n,"version",{start:{line:68,column:35},end:{line:68,column:42}}),n))!=null?o:"")+`',
			WITH_CREDENTIALS: config?.WITH_CREDENTIALS ?? false,
			CREDENTIALS: config?.CREDENTIALS ?? 'include',
			TOKEN: config?.TOKEN,
			USERNAME: config?.USERNAME,
			PASSWORD: config?.PASSWORD,
			HEADERS: config?.HEADERS,
			ENCODE_PATH: config?.ENCODE_PATH,
			interceptors: {
				request: config?.interceptors?.request ?? new Interceptors(),
				response: config?.interceptors?.response ?? new Interceptors(),
      },
		});

`+((o=l(s,"each").call(u,l(n,"services"),{name:"each",hash:{},fn:e.program(14,t,0),inverse:e.noop,data:t,loc:{start:{line:82,column:2},end:{line:84,column:11}}}))!=null?o:"")+`	}
}
`},12:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return "	public readonly "+((o=i(s,"camelCase").call(a,i(n,"name"),{name:"camelCase",hash:{},data:t,loc:{start:{line:60,column:17},end:{line:60,column:37}}}))!=null?o:"")+": "+((o=i(s,"transformServiceName").call(a,i(n,"name"),{name:"transformServiceName",hash:{},data:t,loc:{start:{line:60,column:39},end:{line:60,column:70}}}))!=null?o:"")+`;
`},14:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return "		this."+((o=i(s,"camelCase").call(a,i(n,"name"),{name:"camelCase",hash:{},data:t,loc:{start:{line:83,column:7},end:{line:83,column:27}}}))!=null?o:"")+" = new "+((o=i(s,"transformServiceName").call(a,i(n,"name"),{name:"transformServiceName",hash:{},data:t,loc:{start:{line:83,column:34},end:{line:83,column:65}}}))!=null?o:"")+`(this.request);
`},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return ((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:1,column:0},end:{line:15,column:11}}}))!=null?o:"")+`
`+((o=i(s,"if").call(a,i(n,"services"),{name:"if",hash:{},fn:e.program(5,t,0),inverse:e.noop,data:t,loc:{start:{line:17,column:0},end:{line:21,column:7}}}))!=null?o:"")+`
`+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(8,t,0),inverse:e.program(11,t,0),data:t,loc:{start:{line:23,column:0},end:{line:87,column:11}}}))!=null?o:"")},useData:true};var yn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getHeaders = <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>): Observable<HttpHeaders> => {
	return forkJoin({
		// @ts-ignore
		token: resolve(options, config.TOKEN),
		// @ts-ignore
		username: resolve(options, config.USERNAME),
		// @ts-ignore
		password: resolve(options, config.PASSWORD),
		// @ts-ignore
		additionalHeaders: resolve(options, config.HEADERS),
	}).pipe(
		map(({ token, username, password, additionalHeaders }) => {
			const headers = Object.entries({
				Accept: 'application/json',
				...additionalHeaders,
				...options.headers,
			})
				.filter(([, value]) => value !== undefined && value !== null)
				.reduce((headers, [key, value]) => ({
					...headers,
					[key]: String(value),
				}), {} as Record<string, string>);

			if (isStringWithValue(token)) {
				headers['Authorization'] = \`Bearer \${token}\`;
			}

			if (isStringWithValue(username) && isStringWithValue(password)) {
				const credentials = base64(\`\${username}:\${password}\`);
				headers['Authorization'] = \`Basic \${credentials}\`;
			}

			if (options.body !== undefined) {
				if (options.mediaType) {
					headers['Content-Type'] = options.mediaType;
				} else if (isBlob(options.body)) {
					headers['Content-Type'] = options.body.type || 'application/octet-stream';
				} else if (isString(options.body)) {
					headers['Content-Type'] = 'text/plain';
				} else if (!isFormData(options.body)) {
					headers['Content-Type'] = 'application/json';
				}
			}

			return new HttpHeaders(headers);
		}),
	);
};`},useData:true};var Rn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getRequestBody = (options: ApiRequestOptions): unknown => {
	if (options.body) {
		if (options.mediaType?.includes('application/json') || options.mediaType?.includes('+json')) {
			return JSON.stringify(options.body);
		} else if (isString(options.body) || isBlob(options.body) || isFormData(options.body)) {
			return options.body;
		} else {
			return JSON.stringify(options.body);
		}
	}
	return undefined;
};`},useData:true};var vn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseBody = <T>(response: HttpResponse<T>): T | undefined => {
	if (response.status !== 204 && response.body !== null) {
		return response.body;
	}
	return undefined;
};`},useData:true};var Pn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseHeader = <T>(response: HttpResponse<T>, responseHeader?: string): string | undefined => {
	if (responseHeader) {
		const value = response.headers.get(responseHeader);
		if (isString(value)) {
			return value;
		}
	}
	return undefined;
};`},useData:true};var xn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return `import { HttpClient, HttpHeaders } from '@angular/common/http';
import type { HttpResponse, HttpErrorResponse } from '@angular/common/http';
import { forkJoin, of, throwError } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import type { Observable } from 'rxjs';

import { ApiError } from './ApiError';
import type { ApiRequestOptions } from './ApiRequestOptions';
import type { ApiResult } from './ApiResult';
import type { OpenAPIConfig } from './OpenAPI';

`+((o=e.invokePartial(a(r,"functions/isString"),n,{name:"functions/isString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/isStringWithValue"),n,{name:"functions/isStringWithValue",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/isBlob"),n,{name:"functions/isBlob",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/isFormData"),n,{name:"functions/isFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/base64"),n,{name:"functions/base64",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/getQueryString"),n,{name:"functions/getQueryString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/getUrl"),n,{name:"functions/getUrl",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/getFormData"),n,{name:"functions/getFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/resolve"),n,{name:"functions/resolve",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"angular/getHeaders"),n,{name:"angular/getHeaders",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"angular/getRequestBody"),n,{name:"angular/getRequestBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"angular/sendRequest"),n,{name:"angular/sendRequest",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"angular/getResponseHeader"),n,{name:"angular/getResponseHeader",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"angular/getResponseBody"),n,{name:"angular/getResponseBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(a(r,"functions/catchErrorCodes"),n,{name:"functions/catchErrorCodes",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

/**
 * Request method
 * @param config The OpenAPI configuration object
 * @param http The Angular HTTP client
 * @param options The request options from the service
 * @returns Observable<T>
 * @throws ApiError
 */
export const request = <T>(config: OpenAPIConfig, http: HttpClient, options: ApiRequestOptions<T>): Observable<T> => {
	const url = getUrl(config, options);
	const formData = getFormData(options);
	const body = getRequestBody(options);

	return getHeaders(config, options).pipe(
		switchMap(headers => {
			return sendRequest<T>(config, options, http, url, body, formData, headers);
		}),
		switchMap(async response => {
			for (const fn of config.interceptors.response._fns) {
				response = await fn(response);
			}
			const responseBody = getResponseBody(response);
			const responseHeader = getResponseHeader(response, options.responseHeader);

			let transformedBody = responseBody;
			if (options.responseTransformer && response.ok) {
				transformedBody = await options.responseTransformer(responseBody)
			}

			return {
				url,
				ok: response.ok,
				status: response.status,
				statusText: response.statusText,
				body: responseHeader ?? transformedBody,
			} as ApiResult;
		}),
		catchError((error: HttpErrorResponse) => {
			if (!error.status) {
				return throwError(() => error);
			}
			return of({
				url,
				ok: error.ok,
				status: error.status,
				statusText: error.statusText,
				body: error.error ?? error.statusText,
			} as ApiResult);
		}),
		map(result => {
			catchErrorCodes(options, result);
			return result.body as T;
		}),
		catchError((error: ApiError) => {
			return throwError(() => error);
		}),
	);
};`},usePartial:true,useData:true};var bn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const sendRequest = <T>(
	config: OpenAPIConfig,
	options: ApiRequestOptions<T>,
	http: HttpClient,
	url: string,
	body: unknown,
	formData: FormData | undefined,
	headers: HttpHeaders
): Observable<HttpResponse<T>> => {
	return http.request<T>(options.method, url, {
		headers,
		body: body ?? formData,
		withCredentials: config.WITH_CREDENTIALS,
		observe: 'response',
	});
};`},useData:true};var Cn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `import type { ApiRequestOptions } from './ApiRequestOptions';
import type { ApiResult } from './ApiResult';

export class ApiError extends Error {
	public readonly url: string;
	public readonly status: number;
	public readonly statusText: string;
	public readonly body: unknown;
	public readonly request: ApiRequestOptions;

	constructor(request: ApiRequestOptions, response: ApiResult, message: string) {
		super(message);

		this.name = 'ApiError';
		this.url = response.url;
		this.status = response.status;
		this.statusText = response.statusText;
		this.body = response.body;
		this.request = request;
	}
}`},useData:true};var qn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export type ApiRequestOptions<T = unknown> = {
	readonly body?: any;
	readonly cookies?: Record<string, unknown>;
	readonly errors?: Record<number | string, string>;
	readonly formData?: Record<string, unknown> | any[] | Blob | File;
	readonly headers?: Record<string, unknown>;
	readonly mediaType?: string;
	readonly method:
		| 'DELETE'
		| 'GET'
		| 'HEAD'
		| 'OPTIONS'
		| 'PATCH'
		| 'POST'
		| 'PUT';
	readonly path?: Record<string, unknown>;
	readonly query?: Record<string, unknown>;
	readonly responseHeader?: string;
	readonly responseTransformer?: (data: unknown) => Promise<T>;
	readonly url: string;
};`},useData:true};var An={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export type ApiResult<TData = any> = {
	readonly body: TData;
	readonly ok: boolean;
	readonly status: number;
	readonly statusText: string;
	readonly url: string;
};`},useData:true};var Tn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getHeaders = async <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>): Promise<Record<string, string>> => {
	const [token, username, password, additionalHeaders] = await Promise.all([
		// @ts-ignore
		resolve(options, config.TOKEN),
		// @ts-ignore
		resolve(options, config.USERNAME),
		// @ts-ignore
		resolve(options, config.PASSWORD),
		// @ts-ignore
		resolve(options, config.HEADERS),
	]);

	const headers = Object.entries({
		Accept: 'application/json',
		...additionalHeaders,
		...options.headers,
	})
	.filter(([, value]) => value !== undefined && value !== null)
	.reduce((headers, [key, value]) => ({
		...headers,
		[key]: String(value),
	}), {} as Record<string, string>);

	if (isStringWithValue(token)) {
		headers['Authorization'] = \`Bearer \${token}\`;
	}

	if (isStringWithValue(username) && isStringWithValue(password)) {
		const credentials = base64(\`\${username}:\${password}\`);
		headers['Authorization'] = \`Basic \${credentials}\`;
	}

	if (options.body !== undefined) {
		if (options.mediaType) {
			headers['Content-Type'] = options.mediaType;
		} else if (isBlob(options.body)) {
			headers['Content-Type'] = options.body.type || 'application/octet-stream';
		} else if (isString(options.body)) {
			headers['Content-Type'] = 'text/plain';
		} else if (!isFormData(options.body)) {
			headers['Content-Type'] = 'application/json';
		}
	} else if (options.formData !== undefined) {
		if (options.mediaType) {
			headers['Content-Type'] = options.mediaType;
		}
	}

	return headers;
};`},useData:true};var On={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getRequestBody = (options: ApiRequestOptions): unknown => {
	if (options.body) {
		return options.body;
	}
	return undefined;
};`},useData:true};var wn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseBody = (response: AxiosResponse<unknown>): unknown => {
	if (response.status !== 204) {
		return response.data;
	}
	return undefined;
};`},useData:true};var kn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseHeader = (response: AxiosResponse<unknown>, responseHeader?: string): string | undefined => {
	if (responseHeader) {
		const content = response.headers[responseHeader];
		if (isString(content)) {
			return content;
		}
	}
	return undefined;
};`},useData:true};var En={1:function(e,n,s,r,t){return "ApiResult<T>"},3:function(e,n,s,r,t){return "T"},5:function(e,n,s,r,t){return "result.body"},7:function(e,n,s,r,t){return "result"},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return `import axios from 'axios';
import type { AxiosError, AxiosRequestConfig, AxiosResponse, AxiosInstance } from 'axios';

import { ApiError } from './ApiError';
import type { ApiRequestOptions } from './ApiRequestOptions';
import type { ApiResult } from './ApiResult';
import { CancelablePromise } from './CancelablePromise';
import type { OnCancel } from './CancelablePromise';
import type { OpenAPIConfig } from './OpenAPI';

`+((o=e.invokePartial(i(r,"functions/isString"),n,{name:"functions/isString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isStringWithValue"),n,{name:"functions/isStringWithValue",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isBlob"),n,{name:"functions/isBlob",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isFormData"),n,{name:"functions/isFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isSuccess"),n,{name:"functions/isSuccess",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/base64"),n,{name:"functions/base64",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getQueryString"),n,{name:"functions/getQueryString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getUrl"),n,{name:"functions/getUrl",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getFormData"),n,{name:"functions/getFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/resolve"),n,{name:"functions/resolve",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"axios/getHeaders"),n,{name:"axios/getHeaders",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"axios/getRequestBody"),n,{name:"axios/getRequestBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"axios/sendRequest"),n,{name:"axios/sendRequest",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"axios/getResponseHeader"),n,{name:"axios/getResponseHeader",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"axios/getResponseBody"),n,{name:"axios/getResponseBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/catchErrorCodes"),n,{name:"functions/catchErrorCodes",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

/**
 * Request method
 * @param config The OpenAPI configuration object
 * @param options The request options from the service
 * @param axiosClient The axios client instance to use
 * @returns CancelablePromise<`+((o=i(s,"ifServicesResponse").call(a,"response",{name:"ifServicesResponse",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:64,column:30},end:{line:64,column:108}}}))!=null?o:"")+`>
 * @throws ApiError
 */
export const request = <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>, axiosClient: AxiosInstance = axios): CancelablePromise<`+((o=i(s,"ifServicesResponse").call(a,"response",{name:"ifServicesResponse",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:67,column:136},end:{line:67,column:214}}}))!=null?o:"")+`> => {
	return new CancelablePromise(async (resolve, reject, onCancel) => {
		try {
			const url = getUrl(config, options);
			const formData = getFormData(options);
			const body = getRequestBody(options);
			const headers = await getHeaders(config, options);

			if (!onCancel.isCancelled) {
				let response = await sendRequest<T>(config, options, url, body, formData, headers, onCancel, axiosClient);

				for (const fn of config.interceptors.response._fns) {
					response = await fn(response);
				}

				const responseBody = getResponseBody(response);
				const responseHeader = getResponseHeader(response, options.responseHeader);

				let transformedBody = responseBody;
				if (options.responseTransformer && isSuccess(response.status)) {
					transformedBody = await options.responseTransformer(responseBody)
				}

				const result: ApiResult = {
					url,
					ok: isSuccess(response.status),
					status: response.status,
					statusText: response.statusText,
					body: responseHeader ?? transformedBody,
				};

				catchErrorCodes(options, result);

				resolve(`+((o=i(s,"ifServicesResponse").call(a,"body",{name:"ifServicesResponse",hash:{},fn:e.program(5,t,0),inverse:e.program(7,t,0),data:t,loc:{start:{line:100,column:12},end:{line:100,column:90}}}))!=null?o:"")+`);
			}
		} catch (error) {
			reject(error);
		}
	});
};`},usePartial:true,useData:true};var Sn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const sendRequest = async <T>(
	config: OpenAPIConfig,
	options: ApiRequestOptions<T>,
	url: string,
	body: unknown,
	formData: FormData | undefined,
	headers: Record<string, string>,
	onCancel: OnCancel,
	axiosClient: AxiosInstance
): Promise<AxiosResponse<T>> => {
	const controller = new AbortController();

	let requestConfig: AxiosRequestConfig = {
		data: body ?? formData,
		headers,
		method: options.method,
		signal: controller.signal,
		url,
		withCredentials: config.WITH_CREDENTIALS,
	};

	onCancel(() => controller.abort());

	for (const fn of config.interceptors.request._fns) {
		requestConfig = await fn(requestConfig);
	}

	try {
		return await axiosClient.request(requestConfig);
	} catch (error) {
		const axiosError = error as AxiosError<T>;
		if (axiosError.response) {
			return axiosError.response;
		}
		throw error;
	}
};`},useData:true};var Hn={1:function(e,n,s,r,t){return `import type { HttpClient } from '@angular/common/http';
import type { Observable } from 'rxjs';

import type { ApiRequestOptions } from './ApiRequestOptions';
import type { OpenAPIConfig } from './OpenAPI';
`},3:function(e,n,s,r,t){return `import type { ApiRequestOptions } from './ApiRequestOptions';
import type { CancelablePromise } from './CancelablePromise';
import type { OpenAPIConfig } from './OpenAPI';
`},5:function(e,n,s,r,t){return `	constructor(
		public readonly config: OpenAPIConfig,
		public readonly http: HttpClient,
	) {}
`},7:function(e,n,s,r,t){return `	constructor(public readonly config: OpenAPIConfig) {}
`},9:function(e,n,s,r,t){return `	public abstract request<T>(options: ApiRequestOptions<T>): Observable<T>;
`},11:function(e,n,s,r,t){return `	public abstract request<T>(options: ApiRequestOptions<T>): CancelablePromise<T>;
`},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return ((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:1,column:0},end:{line:11,column:11}}}))!=null?o:"")+`
export abstract class BaseHttpRequest {

`+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(5,t,0),inverse:e.program(7,t,0),data:t,loc:{start:{line:15,column:1},end:{line:22,column:12}}}))!=null?o:"")+`
`+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(9,t,0),inverse:e.program(11,t,0),data:t,loc:{start:{line:24,column:1},end:{line:28,column:12}}}))!=null?o:"")+"}"},useData:true};var Dn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export class CancelError extends Error {
	constructor(message: string) {
		super(message);
		this.name = 'CancelError';
	}

	public get isCancelled(): boolean {
		return true;
	}
}

export interface OnCancel {
	readonly isResolved: boolean;
	readonly isRejected: boolean;
	readonly isCancelled: boolean;

	(cancelHandler: () => void): void;
}

export class CancelablePromise<T> implements Promise<T> {
	private _isResolved: boolean;
	private _isRejected: boolean;
	private _isCancelled: boolean;
	readonly cancelHandlers: (() => void)[];
	readonly promise: Promise<T>;
	private _resolve?: (value: T | PromiseLike<T>) => void;
	private _reject?: (reason?: unknown) => void;

	constructor(
		executor: (
			resolve: (value: T | PromiseLike<T>) => void,
			reject: (reason?: unknown) => void,
			onCancel: OnCancel
		) => void
	) {
		this._isResolved = false;
		this._isRejected = false;
		this._isCancelled = false;
		this.cancelHandlers = [];
		this.promise = new Promise<T>((resolve, reject) => {
			this._resolve = resolve;
			this._reject = reject;

			const onResolve = (value: T | PromiseLike<T>): void => {
				if (this._isResolved || this._isRejected || this._isCancelled) {
					return;
				}
				this._isResolved = true;
				if (this._resolve) this._resolve(value);
			};

			const onReject = (reason?: unknown): void => {
				if (this._isResolved || this._isRejected || this._isCancelled) {
					return;
				}
				this._isRejected = true;
				if (this._reject) this._reject(reason);
			};

			const onCancel = (cancelHandler: () => void): void => {
				if (this._isResolved || this._isRejected || this._isCancelled) {
					return;
				}
				this.cancelHandlers.push(cancelHandler);
			};

			Object.defineProperty(onCancel, 'isResolved', {
				get: (): boolean => this._isResolved,
			});

			Object.defineProperty(onCancel, 'isRejected', {
				get: (): boolean => this._isRejected,
			});

			Object.defineProperty(onCancel, 'isCancelled', {
				get: (): boolean => this._isCancelled,
			});

			return executor(onResolve, onReject, onCancel as OnCancel);
		});
	}

	get [Symbol.toStringTag]() {
		return "Cancellable Promise";
	}

	public then<TResult1 = T, TResult2 = never>(
		onFulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | null,
		onRejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null
	): Promise<TResult1 | TResult2> {
		return this.promise.then(onFulfilled, onRejected);
	}

	public catch<TResult = never>(
		onRejected?: ((reason: unknown) => TResult | PromiseLike<TResult>) | null
	): Promise<T | TResult> {
		return this.promise.catch(onRejected);
	}

	public finally(onFinally?: (() => void) | null): Promise<T> {
		return this.promise.finally(onFinally);
	}

	public cancel(): void {
		if (this._isResolved || this._isRejected || this._isCancelled) {
			return;
		}
		this._isCancelled = true;
		if (this.cancelHandlers.length) {
			try {
				for (const cancelHandler of this.cancelHandlers) {
					cancelHandler();
				}
			} catch (error) {
				console.warn('Cancellation threw an error', error);
				return;
			}
		}
		this.cancelHandlers.length = 0;
		if (this._reject) this._reject(new CancelError('Request aborted'));
	}

	public get isCancelled(): boolean {
		return this._isCancelled;
	}
}`},useData:true};var In={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getHeaders = async <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>): Promise<Headers> => {
	const [token, username, password, additionalHeaders] = await Promise.all([
		// @ts-ignore
		resolve(options, config.TOKEN),
		// @ts-ignore
		resolve(options, config.USERNAME),
		// @ts-ignore
		resolve(options, config.PASSWORD),
		// @ts-ignore
		resolve(options, config.HEADERS),
	]);

	const headers = Object.entries({
		Accept: 'application/json',
		...additionalHeaders,
		...options.headers,
	})
		.filter(([, value]) => value !== undefined && value !== null)
		.reduce((headers, [key, value]) => ({
			...headers,
			[key]: String(value),
		}), {} as Record<string, string>);

	if (isStringWithValue(token)) {
		headers['Authorization'] = \`Bearer \${token}\`;
	}

	if (isStringWithValue(username) && isStringWithValue(password)) {
		const credentials = base64(\`\${username}:\${password}\`);
		headers['Authorization'] = \`Basic \${credentials}\`;
	}

	if (options.body !== undefined) {
		if (options.mediaType) {
			headers['Content-Type'] = options.mediaType;
		} else if (isBlob(options.body)) {
			headers['Content-Type'] = options.body.type || 'application/octet-stream';
		} else if (isString(options.body)) {
			headers['Content-Type'] = 'text/plain';
		} else if (!isFormData(options.body)) {
			headers['Content-Type'] = 'application/json';
		}
	}

	return new Headers(headers);
};`},useData:true};var Bn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getRequestBody = (options: ApiRequestOptions): unknown => {
	if (options.body !== undefined) {
		if (options.mediaType?.includes('application/json') || options.mediaType?.includes('+json')) {
			return JSON.stringify(options.body);
		} else if (isString(options.body) || isBlob(options.body) || isFormData(options.body)) {
			return options.body;
		} else {
			return JSON.stringify(options.body);
		}
	}
	return undefined;
};`},useData:true};var $n={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseBody = async (response: Response): Promise<unknown> => {
	if (response.status !== 204) {
		try {
			const contentType = response.headers.get('Content-Type');
			if (contentType) {
				const binaryTypes = ['application/octet-stream', 'application/pdf', 'application/zip', 'audio/', 'image/', 'video/'];
				if (contentType.includes('application/json') || contentType.includes('+json')) {
					return await response.json();
				} else if (binaryTypes.some(type => contentType.includes(type))) {
					return await response.blob();
				} else if (contentType.includes('multipart/form-data')) {
					return await response.formData();
				} else if (contentType.includes('text/')) {
					return await response.text();
				}
			}
		} catch (error) {
			console.error(error);
		}
	}
	return undefined;
};`},useData:true};var jn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseHeader = (response: Response, responseHeader?: string): string | undefined => {
	if (responseHeader) {
		const content = response.headers.get(responseHeader);
		if (isString(content)) {
			return content;
		}
	}
	return undefined;
};`},useData:true};var Nn={1:function(e,n,s,r,t){return `import fetch, { FormData, Headers } from 'node-fetch';
import type { RequestInit, Response } from 'node-fetch';

`},3:function(e,n,s,r,t){return "ApiResult<T>"},5:function(e,n,s,r,t){return "T"},7:function(e,n,s,r,t){return "result.body"},9:function(e,n,s,r,t){return "result"},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return ((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/node",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.noop,data:t,loc:{start:{line:1,column:0},end:{line:5,column:11}}}))!=null?o:"")+`import { ApiError } from './ApiError';
import type { ApiRequestOptions } from './ApiRequestOptions';
import type { ApiResult } from './ApiResult';
import { CancelablePromise } from './CancelablePromise';
import type { OnCancel } from './CancelablePromise';
import type { OpenAPIConfig } from './OpenAPI';

`+((o=e.invokePartial(i(r,"functions/isString"),n,{name:"functions/isString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isStringWithValue"),n,{name:"functions/isStringWithValue",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isBlob"),n,{name:"functions/isBlob",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isFormData"),n,{name:"functions/isFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/base64"),n,{name:"functions/base64",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getQueryString"),n,{name:"functions/getQueryString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getUrl"),n,{name:"functions/getUrl",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getFormData"),n,{name:"functions/getFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/resolve"),n,{name:"functions/resolve",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"fetch/getHeaders"),n,{name:"fetch/getHeaders",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"fetch/getRequestBody"),n,{name:"fetch/getRequestBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"fetch/sendRequest"),n,{name:"fetch/sendRequest",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"fetch/getResponseHeader"),n,{name:"fetch/getResponseHeader",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"fetch/getResponseBody"),n,{name:"fetch/getResponseBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/catchErrorCodes"),n,{name:"functions/catchErrorCodes",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

/**
 * Request method
 * @param config The OpenAPI configuration object
 * @param options The request options from the service
 * @returns CancelablePromise<`+((o=i(s,"ifServicesResponse").call(a,"response",{name:"ifServicesResponse",hash:{},fn:e.program(3,t,0),inverse:e.program(5,t,0),data:t,loc:{start:{line:62,column:30},end:{line:62,column:108}}}))!=null?o:"")+`>
 * @throws ApiError
 */
export const request = <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>): CancelablePromise<`+((o=i(s,"ifServicesResponse").call(a,"response",{name:"ifServicesResponse",hash:{},fn:e.program(3,t,0),inverse:e.program(5,t,0),data:t,loc:{start:{line:65,column:100},end:{line:65,column:178}}}))!=null?o:"")+`> => {
	return new CancelablePromise(async (resolve, reject, onCancel) => {
		try {
			const url = getUrl(config, options);
			const formData = getFormData(options);
			const body = getRequestBody(options);
			const headers = await getHeaders(config, options);

			if (!onCancel.isCancelled) {
				let response = await sendRequest(config, options, url, body, formData, headers, onCancel);

				for (const fn of config.interceptors.response._fns) {
					response = await fn(response);
				}

				const responseBody = await getResponseBody(response);
				const responseHeader = getResponseHeader(response, options.responseHeader);

				let transformedBody = responseBody;
				if (options.responseTransformer && response.ok) {
					transformedBody = await options.responseTransformer(responseBody)
				}

				const result: ApiResult = {
					url,
					ok: response.ok,
					status: response.status,
					statusText: response.statusText,
					body: responseHeader ?? transformedBody,
				};

				catchErrorCodes(options, result);

				resolve(`+((o=i(s,"ifServicesResponse").call(a,"body",{name:"ifServicesResponse",hash:{},fn:e.program(7,t,0),inverse:e.program(9,t,0),data:t,loc:{start:{line:98,column:12},end:{line:98,column:90}}}))!=null?o:"")+`);
			}
		} catch (error) {
			reject(error);
		}
	});
};`},usePartial:true,useData:true};var _n={1:function(e,n,s,r,t){return `	if (config.WITH_CREDENTIALS) {
		request.credentials = config.CREDENTIALS;
	}
`},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return `export const sendRequest = async (
	config: OpenAPIConfig,
	options: ApiRequestOptions,
	url: string,
	body: any,
	formData: FormData | undefined,
	headers: Headers,
	onCancel: OnCancel
): Promise<Response> => {
	const controller = new AbortController();

	let request: RequestInit = {
		headers,
		body: body ?? formData,
		method: options.method,
		signal: controller.signal,
	};

`+((o=a(s,"equals").call(n??(e.nullContext||{}),a(a(a(a(t,"root"),"$config"),"client"),"name"),"legacy/fetch",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.noop,data:t,loc:{start:{line:19,column:1},end:{line:23,column:12}}}))!=null?o:"")+`
	for (const fn of config.interceptors.request._fns) {
		request = await fn(request);
	}

	onCancel(() => controller.abort());

	return await fetch(url, request);
};`},useData:true};var Fn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const base64 = (str: string): string => {
	try {
		return btoa(str);
	} catch (err) {
		// @ts-ignore
		return Buffer.from(str).toString('base64');
	}
};`},useData:true};var Ln={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const catchErrorCodes = (options: ApiRequestOptions, result: ApiResult): void => {
	const errors: Record<number, string> = {
		400: 'Bad Request',
		401: 'Unauthorized',
		402: 'Payment Required',
		403: 'Forbidden',
		404: 'Not Found',
		405: 'Method Not Allowed',
		406: 'Not Acceptable',
		407: 'Proxy Authentication Required',
		408: 'Request Timeout',
		409: 'Conflict',
		410: 'Gone',
		411: 'Length Required',
		412: 'Precondition Failed',
		413: 'Payload Too Large',
		414: 'URI Too Long',
		415: 'Unsupported Media Type',
		416: 'Range Not Satisfiable',
		417: 'Expectation Failed',
		418: 'Im a teapot',
		421: 'Misdirected Request',
		422: 'Unprocessable Content',
		423: 'Locked',
		424: 'Failed Dependency',
		425: 'Too Early',
		426: 'Upgrade Required',
		428: 'Precondition Required',
		429: 'Too Many Requests',
		431: 'Request Header Fields Too Large',
		451: 'Unavailable For Legal Reasons',
		500: 'Internal Server Error',
		501: 'Not Implemented',
		502: 'Bad Gateway',
		503: 'Service Unavailable',
		504: 'Gateway Timeout',
		505: 'HTTP Version Not Supported',
		506: 'Variant Also Negotiates',
		507: 'Insufficient Storage',
		508: 'Loop Detected',
		510: 'Not Extended',
		511: 'Network Authentication Required',
		...options.errors,
	}

	const error = errors[result.status];
	if (error) {
		throw new ApiError(options, result, error);
	}

	if (!result.ok) {
		const errorStatus = result.status ?? 'unknown';
		const errorStatusText = result.statusText ?? 'unknown';
		const errorBody = (() => {
			try {
				return JSON.stringify(result.body, null, 2);
			} catch (e) {
				return undefined;
			}
		})();

		throw new ApiError(options, result,
			\`Generic Error: status: \${errorStatus}; status text: \${errorStatusText}; body: \${errorBody}\`
		);
	}
};`},useData:true};var Wn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getFormData = (options: ApiRequestOptions): FormData | undefined => {
	if (options.formData) {
		const formData = new FormData();

		const process = (key: string, value: unknown) => {
			if (isString(value) || isBlob(value)) {
				formData.append(key, value);
			} else {
				formData.append(key, JSON.stringify(value));
			}
		};

		Object.entries(options.formData)
			.filter(([, value]) => value !== undefined && value !== null)
			.forEach(([key, value]) => {
				if (Array.isArray(value)) {
					value.forEach(v => process(key, v));
				} else {
					process(key, value);
				}
			});

		return formData;
	}
	return undefined;
};`},useData:true};var Mn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getQueryString = (params: Record<string, unknown>): string => {
	const qs: string[] = [];

	const append = (key: string, value: unknown) => {
		qs.push(\`\${encodeURIComponent(key)}=\${encodeURIComponent(String(value))}\`);
	};

	const encodePair = (key: string, value: unknown) => {
		if (value === undefined || value === null) {
			return;
		}

		if (value instanceof Date) {
			append(key, value.toISOString());
		} else if (Array.isArray(value)) {
			value.forEach(v => encodePair(key, v));
		} else if (typeof value === 'object') {
			Object.entries(value).forEach(([k, v]) => encodePair(\`\${key}[\${k}]\`, v));
		} else {
			append(key, value);
		}
	};

	Object.entries(params).forEach(([key, value]) => encodePair(key, value));

	return qs.length ? \`?\${qs.join('&')}\` : '';
};`},useData:true};var Un={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `const getUrl = (config: OpenAPIConfig, options: ApiRequestOptions): string => {
	const encoder = config.ENCODE_PATH || encodeURI;

	const path = options.url
		.replace('{api-version}', config.VERSION)
		.replace(/{(.*?)}/g, (substring: string, group: string) => {
			if (options.path?.hasOwnProperty(group)) {
				return encoder(String(options.path[group]));
			}
			return substring;
		});

	const url = config.BASE + path;
	return options.query ? url + getQueryString(options.query) : url;
};`},useData:true};var Vn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const isBlob = (value: any): value is Blob => {
	return value instanceof Blob;
};`},useData:true};var Gn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const isFormData = (value: unknown): value is FormData => {
	return value instanceof FormData;
};`},useData:true};var zn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const isString = (value: unknown): value is string => {
	return typeof value === 'string';
};`},useData:true};var Kn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const isStringWithValue = (value: unknown): value is string => {
	return isString(value) && value !== '';
};`},useData:true};var Jn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const isSuccess = (status: number): boolean => {
	return status >= 200 && status < 300;
};`},useData:true};var Qn={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `type Resolver<T> = (options: ApiRequestOptions<T>) => Promise<T>;

export const resolve = async <T>(options: ApiRequestOptions<T>, resolver?: T | Resolver<T>): Promise<T | undefined> => {
	if (typeof resolver === 'function') {
		return (resolver as Resolver<T>)(options);
	}
	return resolver;
};`},useData:true};var Xn={1:function(e,n,s,r,t){return `import { Inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import type { Observable } from 'rxjs';

import type { ApiRequestOptions } from './ApiRequestOptions';
import { BaseHttpRequest } from './BaseHttpRequest';
import type { OpenAPIConfig } from './OpenAPI';
import { OpenAPI } from './OpenAPI';
import { request as __request } from './request';
`},3:function(e,n,s,r,t){return `import type { ApiRequestOptions } from './ApiRequestOptions';
import { BaseHttpRequest } from './BaseHttpRequest';
import type { CancelablePromise } from './CancelablePromise';
import type { OpenAPIConfig } from './OpenAPI';
import { request as __request } from './request';
`},5:function(e,n,s,r,t){return `@Injectable()
`},7:function(e,n,s,r,t){return `	constructor(
		@Inject(OpenAPI)
		config: OpenAPIConfig,
		http: HttpClient,
	) {
		super(config, http);
	}
`},9:function(e,n,s,r,t){return `	constructor(config: OpenAPIConfig) {
		super(config);
	}
`},11:function(e,n,s,r,t){return `	/**
	 * Request method
	 * @param options The request options from the service
	 * @returns Observable<T>
	 * @throws ApiError
	 */
	public override request<T>(options: ApiRequestOptions<T>): Observable<T> {
		return __request(this.config, this.http, options);
	}
`},13:function(e,n,s,r,t){return `	/**
	 * Request method
	 * @param options The request options from the service
	 * @returns CancelablePromise<T>
	 * @throws ApiError
	 */
	public override request<T>(options: ApiRequestOptions<T>): CancelablePromise<T> {
		return __request(this.config, options);
	}
`},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return ((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:1,column:0},end:{line:17,column:11}}}))!=null?o:"")+`
`+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(5,t,0),inverse:e.noop,data:t,loc:{start:{line:19,column:0},end:{line:21,column:11}}}))!=null?o:"")+"export class "+((o=e.lambda(e.strict(n,"httpRequest",{start:{line:22,column:15},end:{line:22,column:26}}),n))!=null?o:"")+` extends BaseHttpRequest {

`+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(7,t,0),inverse:e.program(9,t,0),data:t,loc:{start:{line:24,column:1},end:{line:36,column:12}}}))!=null?o:"")+`
`+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(11,t,0),inverse:e.program(13,t,0),data:t,loc:{start:{line:38,column:1},end:{line:58,column:12}}}))!=null?o:"")+"}"},useData:true};var Yn={1:function(e,n,s,r,t){return `import type { HttpResponse } from '@angular/common/http';
`},3:function(e,n,s,r,t){return `import type { AxiosRequestConfig, AxiosResponse } from 'axios';
`},5:function(e,n,s,r,t){return `import type { RequestInit, Response } from 'node-fetch';
`},7:function(e,n,s,r,t){return `		response: Interceptors<HttpResponse<any>>;
`},9:function(e,n,s,r,t){return `		request: Interceptors<AxiosRequestConfig>;
		response: Interceptors<AxiosResponse>;
`},11:function(e,n,s,r,t){return `		request: Interceptors<RequestInit>;
		response: Interceptors<Response>;
`},13:function(e,n,s,r,t){return `		request: Interceptors<XMLHttpRequest>;
		response: Interceptors<XMLHttpRequest>;
`},15:function(e,n,s,r,t){return `		request: new Interceptors(),
`},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.strict,u=e.lambda,l=e.lookupProperty||function(c,m){if(Object.prototype.hasOwnProperty.call(c,m))return c[m]};return ((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.noop,data:t,loc:{start:{line:1,column:0},end:{line:3,column:11}}}))!=null?o:"")+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/axios",{name:"equals",hash:{},fn:e.program(3,t,0),inverse:e.noop,data:t,loc:{start:{line:4,column:0},end:{line:6,column:11}}}))!=null?o:"")+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/node",{name:"equals",hash:{},fn:e.program(5,t,0),inverse:e.noop,data:t,loc:{start:{line:7,column:0},end:{line:9,column:11}}}))!=null?o:"")+`import type { ApiRequestOptions } from './ApiRequestOptions';

type Headers = Record<string, string>;
type Middleware<T> = (value: T) => T | Promise<T>;
type Resolver<T> = (options: ApiRequestOptions<T>) => Promise<T>;

export class Interceptors<T> {
  _fns: Middleware<T>[];

  constructor() {
    this._fns = [];
  }

  eject(fn: Middleware<T>): void {
    const index = this._fns.indexOf(fn);
    if (index !== -1) {
      this._fns = [...this._fns.slice(0, index), ...this._fns.slice(index + 1)];
    }
  }

  use(fn: Middleware<T>): void {
    this._fns = [...this._fns, fn];
  }
}

export type OpenAPIConfig = {
	BASE: string;
	CREDENTIALS: 'include' | 'omit' | 'same-origin';
	ENCODE_PATH?: ((path: string) => string) | undefined;
	HEADERS?: Headers | Resolver<Headers> | undefined;
	PASSWORD?: string | Resolver<string> | undefined;
	TOKEN?: string | Resolver<string> | undefined;
	USERNAME?: string | Resolver<string> | undefined;
	VERSION: string;
	WITH_CREDENTIALS: boolean;
	interceptors: {
`+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(7,t,0),inverse:e.noop,data:t,loc:{start:{line:46,column:2},end:{line:48,column:13}}}))!=null?o:"")+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/axios",{name:"equals",hash:{},fn:e.program(9,t,0),inverse:e.noop,data:t,loc:{start:{line:49,column:2},end:{line:52,column:13}}}))!=null?o:"")+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/fetch",{name:"equals",hash:{},fn:e.program(11,t,0),inverse:e.noop,data:t,loc:{start:{line:53,column:2},end:{line:56,column:13}}}))!=null?o:"")+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/node",{name:"equals",hash:{},fn:e.program(11,t,0),inverse:e.noop,data:t,loc:{start:{line:57,column:2},end:{line:60,column:13}}}))!=null?o:"")+((o=l(s,"equals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/xhr",{name:"equals",hash:{},fn:e.program(13,t,0),inverse:e.noop,data:t,loc:{start:{line:61,column:2},end:{line:64,column:13}}}))!=null?o:"")+`	};
};

export const OpenAPI: OpenAPIConfig = {
	BASE: '`+((o=u(i(n,"server",{start:{line:69,column:11},end:{line:69,column:17}}),n))!=null?o:"")+`',
	CREDENTIALS: 'include',
	ENCODE_PATH: undefined,
	HEADERS: undefined,
	PASSWORD: undefined,
	TOKEN: undefined,
	USERNAME: undefined,
	VERSION: '`+((o=u(i(n,"version",{start:{line:76,column:14},end:{line:76,column:21}}),n))!=null?o:"")+`',
	WITH_CREDENTIALS: false,
	interceptors: {
`+((o=l(s,"notEquals").call(a,l(l(l(l(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"notEquals",hash:{},fn:e.program(15,t,0),inverse:e.noop,data:t,loc:{start:{line:79,column:2},end:{line:81,column:16}}}))!=null?o:"")+`		response: new Interceptors(),
	},
};`},useData:true};var Zn={1:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return (o=e.invokePartial(a(r,"angular/request"),n,{name:"angular/request",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:""},3:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return (o=e.invokePartial(a(r,"axios/request"),n,{name:"axios/request",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:""},5:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return (o=e.invokePartial(a(r,"fetch/request"),n,{name:"fetch/request",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:""},7:function(e,n,s,r,t){var o,a=e.lookupProperty||function(i,u){if(Object.prototype.hasOwnProperty.call(i,u))return i[u]};return (o=e.invokePartial(a(r,"xhr/request"),n,{name:"xhr/request",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:""},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return ((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/angular",{name:"equals",hash:{},fn:e.program(1,t,0),inverse:e.noop,data:t,loc:{start:{line:1,column:0},end:{line:1,column:87}}}))!=null?o:"")+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/axios",{name:"equals",hash:{},fn:e.program(3,t,0),inverse:e.noop,data:t,loc:{start:{line:2,column:0},end:{line:2,column:83}}}))!=null?o:"")+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/fetch",{name:"equals",hash:{},fn:e.program(5,t,0),inverse:e.noop,data:t,loc:{start:{line:3,column:0},end:{line:3,column:83}}}))!=null?o:"")+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/node",{name:"equals",hash:{},fn:e.program(5,t,0),inverse:e.noop,data:t,loc:{start:{line:4,column:0},end:{line:4,column:82}}}))!=null?o:"")+((o=i(s,"equals").call(a,i(i(i(i(t,"root"),"$config"),"client"),"name"),"legacy/xhr",{name:"equals",hash:{},fn:e.program(7,t,0),inverse:e.noop,data:t,loc:{start:{line:5,column:0},end:{line:5,column:79}}}))!=null?o:"")},usePartial:true,useData:true};var et={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getHeaders = async <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>): Promise<Headers> => {
	const [token, username, password, additionalHeaders] = await Promise.all([
		// @ts-ignore
		resolve(options, config.TOKEN),
		// @ts-ignore
		resolve(options, config.USERNAME),
		// @ts-ignore
		resolve(options, config.PASSWORD),
		// @ts-ignore
		resolve(options, config.HEADERS),
	]);

	const headers = Object.entries({
		Accept: 'application/json',
		...additionalHeaders,
		...options.headers,
	})
		.filter(([, value]) => value !== undefined && value !== null)
		.reduce((headers, [key, value]) => ({
			...headers,
			[key]: String(value),
		}), {} as Record<string, string>);

	if (isStringWithValue(token)) {
		headers['Authorization'] = \`Bearer \${token}\`;
	}

	if (isStringWithValue(username) && isStringWithValue(password)) {
		const credentials = base64(\`\${username}:\${password}\`);
		headers['Authorization'] = \`Basic \${credentials}\`;
	}

	if (options.body !== undefined) {
		if (options.mediaType) {
			headers['Content-Type'] = options.mediaType;
		} else if (isBlob(options.body)) {
			headers['Content-Type'] = options.body.type || 'application/octet-stream';
		} else if (isString(options.body)) {
			headers['Content-Type'] = 'text/plain';
		} else if (!isFormData(options.body)) {
			headers['Content-Type'] = 'application/json';
		}
	}

	return new Headers(headers);
};`},useData:true};var nt={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getRequestBody = (options: ApiRequestOptions): unknown => {
	if (options.body !== undefined) {
		if (options.mediaType?.includes('application/json') || options.mediaType?.includes('+json')) {
			return JSON.stringify(options.body);
		} else if (isString(options.body) || isBlob(options.body) || isFormData(options.body)) {
			return options.body;
		} else {
			return JSON.stringify(options.body);
		}
	}
	return undefined;
};`},useData:true};var tt={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseBody = (xhr: XMLHttpRequest): unknown => {
	if (xhr.status !== 204) {
		try {
			const contentType = xhr.getResponseHeader('Content-Type');
			if (contentType) {
				if (contentType.includes('application/json') || contentType.includes('+json')) {
					return JSON.parse(xhr.responseText);
				} else {
					return xhr.responseText;
				}
			}
		} catch (error) {
			console.error(error);
		}
	}
	return undefined;
};`},useData:true};var rt={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const getResponseHeader = (xhr: XMLHttpRequest, responseHeader?: string): string | undefined => {
	if (responseHeader) {
		const content = xhr.getResponseHeader(responseHeader);
		if (isString(content)) {
			return content;
		}
	}
	return undefined;
};`},useData:true};var ot={1:function(e,n,s,r,t){return "ApiResult<T>"},3:function(e,n,s,r,t){return "T"},5:function(e,n,s,r,t){return "result.body"},7:function(e,n,s,r,t){return "result"},compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){var o,a=n??(e.nullContext||{}),i=e.lookupProperty||function(u,l){if(Object.prototype.hasOwnProperty.call(u,l))return u[l]};return `import { ApiError } from './ApiError';
import type { ApiRequestOptions } from './ApiRequestOptions';
import type { ApiResult } from './ApiResult';
import { CancelablePromise } from './CancelablePromise';
import type { OnCancel } from './CancelablePromise';
import type { OpenAPIConfig } from './OpenAPI';

`+((o=e.invokePartial(i(r,"functions/isString"),n,{name:"functions/isString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isStringWithValue"),n,{name:"functions/isStringWithValue",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isBlob"),n,{name:"functions/isBlob",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isFormData"),n,{name:"functions/isFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/isSuccess"),n,{name:"functions/isSuccess",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/base64"),n,{name:"functions/base64",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getQueryString"),n,{name:"functions/getQueryString",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getUrl"),n,{name:"functions/getUrl",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/getFormData"),n,{name:"functions/getFormData",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/resolve"),n,{name:"functions/resolve",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"fetch/getHeaders"),n,{name:"fetch/getHeaders",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"xhr/getRequestBody"),n,{name:"xhr/getRequestBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"xhr/sendRequest"),n,{name:"xhr/sendRequest",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"xhr/getResponseHeader"),n,{name:"xhr/getResponseHeader",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"xhr/getResponseBody"),n,{name:"xhr/getResponseBody",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

`+((o=e.invokePartial(i(r,"functions/catchErrorCodes"),n,{name:"functions/catchErrorCodes",data:t,helpers:s,partials:r,decorators:e.decorators}))!=null?o:"")+`

/**
 * Request method
 * @param config The OpenAPI configuration object
 * @param options The request options from the service
 * @returns CancelablePromise<`+((o=i(s,"ifServicesResponse").call(a,"response",{name:"ifServicesResponse",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:60,column:30},end:{line:60,column:108}}}))!=null?o:"")+`>
 * @throws ApiError
 */
export const request = <T>(config: OpenAPIConfig, options: ApiRequestOptions<T>): CancelablePromise<`+((o=i(s,"ifServicesResponse").call(a,"response",{name:"ifServicesResponse",hash:{},fn:e.program(1,t,0),inverse:e.program(3,t,0),data:t,loc:{start:{line:63,column:100},end:{line:63,column:178}}}))!=null?o:"")+`> => {
	return new CancelablePromise(async (resolve, reject, onCancel) => {
		try {
			const url = getUrl(config, options);
			const formData = getFormData(options);
			const body = getRequestBody(options);
			const headers = await getHeaders(config, options);

			if (!onCancel.isCancelled) {
				let response = await sendRequest(config, options, url, body, formData, headers, onCancel);

				for (const fn of config.interceptors.response._fns) {
					response = await fn(response);
				}

				const responseBody = getResponseBody(response);
				const responseHeader = getResponseHeader(response, options.responseHeader);

				let transformedBody = responseBody;
				if (options.responseTransformer && isSuccess(response.status)) {
					transformedBody = await options.responseTransformer(responseBody)
				}

				const result: ApiResult = {
					url,
					ok: isSuccess(response.status),
					status: response.status,
					statusText: response.statusText,
					body: responseHeader ?? transformedBody,
				};

				catchErrorCodes(options, result);

				resolve(`+((o=i(s,"ifServicesResponse").call(a,"body",{name:"ifServicesResponse",hash:{},fn:e.program(5,t,0),inverse:e.program(7,t,0),data:t,loc:{start:{line:96,column:12},end:{line:96,column:90}}}))!=null?o:"")+`);
			}
		} catch (error) {
			reject(error);
		}
	});
};`},usePartial:true,useData:true};var st={compiler:[8,">= 4.3.0"],main:function(e,n,s,r,t){return `export const sendRequest = async (
	config: OpenAPIConfig,
	options: ApiRequestOptions,
	url: string,
	body: any,
	formData: FormData | undefined,
	headers: Headers,
	onCancel: OnCancel
): Promise<XMLHttpRequest> => {
	let xhr = new XMLHttpRequest();
	xhr.open(options.method, url, true);
	xhr.withCredentials = config.WITH_CREDENTIALS;

	headers.forEach((value, key) => {
		xhr.setRequestHeader(key, value);
	});

	return new Promise<XMLHttpRequest>(async (resolve, reject) => {
		xhr.onload = () => resolve(xhr);
		xhr.onabort = () => reject(new Error('Request aborted'));
		xhr.onerror = () => reject(new Error('Network error'));

		for (const fn of config.interceptors.request._fns) {
			xhr = await fn(xhr);
		}

		xhr.send(body ?? formData);

		onCancel(() => xhr.abort());
	});
};`},useData:true};var Kt=()=>{p.registerHelper("camelCase",function(e){return m({case:"camelCase",value:e})}),p.registerHelper("equals",function(e,n,s){return e===n?s.fn(this):s.inverse(this)}),p.registerHelper("ifServicesResponse",function(e$1,n){return e().plugins["@hey-api/sdk"]?.response===e$1?n.fn(this):n.inverse(this)}),p.registerHelper("ifdef",function(...e){let n=e.pop();return e.every(s=>!s)?n.inverse(this):n.fn(this)}),p.registerHelper("notEquals",function(e,n,s){return e!==n?s.fn(this):s.inverse(this)}),p.registerHelper("transformServiceName",function(e$1){return p$1({config:e(),name:e$1})});},it=()=>{Kt();let e={client:p.template(hn),core:{apiError:p.template(Cn),apiRequestOptions:p.template(qn),apiResult:p.template(An),baseHttpRequest:p.template(Hn),cancelablePromise:p.template(Dn),httpRequest:p.template(Xn),request:p.template(Zn),settings:p.template(Yn)}};return p.registerPartial("functions/base64",p.template(Fn)),p.registerPartial("functions/catchErrorCodes",p.template(Ln)),p.registerPartial("functions/getFormData",p.template(Wn)),p.registerPartial("functions/getQueryString",p.template(Mn)),p.registerPartial("functions/getUrl",p.template(Un)),p.registerPartial("functions/isBlob",p.template(Vn)),p.registerPartial("functions/isFormData",p.template(Gn)),p.registerPartial("functions/isString",p.template(zn)),p.registerPartial("functions/isStringWithValue",p.template(Kn)),p.registerPartial("functions/isSuccess",p.template(Jn)),p.registerPartial("functions/resolve",p.template(Qn)),p.registerPartial("fetch/getHeaders",p.template(In)),p.registerPartial("fetch/getRequestBody",p.template(Bn)),p.registerPartial("fetch/getResponseBody",p.template($n)),p.registerPartial("fetch/getResponseHeader",p.template(jn)),p.registerPartial("fetch/request",p.template(Nn)),p.registerPartial("fetch/sendRequest",p.template(_n)),p.registerPartial("xhr/getHeaders",p.template(et)),p.registerPartial("xhr/getRequestBody",p.template(nt)),p.registerPartial("xhr/getResponseBody",p.template(tt)),p.registerPartial("xhr/getResponseHeader",p.template(rt)),p.registerPartial("xhr/request",p.template(ot)),p.registerPartial("xhr/sendRequest",p.template(st)),p.registerPartial("axios/getHeaders",p.template(Tn)),p.registerPartial("axios/getRequestBody",p.template(On)),p.registerPartial("axios/getResponseBody",p.template(wn)),p.registerPartial("axios/getResponseHeader",p.template(kn)),p.registerPartial("axios/request",p.template(En)),p.registerPartial("axios/sendRequest",p.template(Sn)),p.registerPartial("angular/getHeaders",p.template(yn)),p.registerPartial("angular/getRequestBody",p.template(Rn)),p.registerPartial("angular/getResponseBody",p.template(vn)),p.registerPartial("angular/getResponseHeader",p.template(Pn)),p.registerPartial("angular/request",p.template(xn)),p.registerPartial("angular/sendRequest",p.template(bn)),e};var Jt={stringCase:m};var oi=async e=>{let n=typeof e=="function"?await e():e,s=[];try{f.start("createClient"),f.start("config"),s=await x(n),f.end("config"),f.start("handlebars");let r=it();f.end("handlebars");let o=(await Promise.all(s.map(i=>re({config:i,templates:r})))).filter(i=>!!i);f.end("createClient");let a=s[0];return a&&a.logs.level==="debug"&&new _({totalMark:"createClient"}).report({marks:["config","openapi","handlebars","parser","generator","postprocess"]}),o}catch(r){let t$1=s[0];if(!(t$1?t$1.dryRun:n?.dryRun)){let a=t$1?.logs??t(n);if(a.level!=="silent"&&a.file){let i=`openapi-ts-error-${Date.now()}.log`,u=R.resolve(process.cwd(),a.path??"");j(u);let l=R.resolve(u,i);y.writeFileSync(l,`${r.message}
${r.stack}`),console.error(`\u{1F525} Unexpected error occurred. Log saved to ${l}`);}}throw console.error(`\u{1F525} Unexpected error occurred. ${r.message}`),r}},si=async e=>typeof e=="function"?await e():e;export{oi as createClient,si as defineConfig,Jt as utils};//# sourceMappingURL=index.js.map
