// thirdweb/dist/cjs/wallets/in-app/core/authentication/backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.backendAuthenticate = backendAuthenticate;
const fetch_js_1 = require("../../../../utils/fetch.js");
const json_js_1 = require("../../../../utils/json.js");
const getLoginPath_js_1 = require("./getLoginPath.js");
/**
 * Authenticates via the wallet secret
 * @internal
 */
async function backendAuthenticate(args) {
    const clientFetch = (0, fetch_js_1.getClientFetch)(args.client, args.ecosystem);
    const path = (0, getLoginPath_js_1.getLoginUrl)({
        authOption: "backend",
        client: args.client,
        ecosystem: args.ecosystem,
    });
    const res = await clientFetch(`${path}`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: (0, json_js_1.stringify)({
            walletSecret: args.walletSecret,
        }),
    });
    if (!res.ok) {
        throw new Error("Failed to generate backend account");
    }
    return (await res.json());
}
