// thirdweb/dist/cjs/extensions/erc20/drops/read/canClaim.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.canClaim = canClaim;
const extract_error_js_1 = require("../../../../transaction/extract-error.js");
const get_claim_params_js_1 = require("../../../../utils/extensions/drops/get-claim-params.js");
const verifyClaim_js_1 = require("../../__generated__/DropERC20/read/verifyClaim.js");
const getActiveClaimConditionId_js_1 = require("../../__generated__/IDropERC20/read/getActiveClaimConditionId.js");
const decimals_js_1 = require("../../read/decimals.js");
/**
 * Check if a user can claim a drop.
 *
 * @param options - The options for the transaction.
 * @returns Whether the user can claim the drop.
 *
 * @example
 * ```ts
 * const claimResult = await canClaim({
 *   contract: contract,
 *   claimer: "0x1234567890123456789012345678901234567890",
 *   quantity: "1",
 * });
 * ```
 *
 * @extension ERC20
 */
async function canClaim(options) {
    const quantityWei = await (async () => {
        if ("quantityInWei" in options) {
            return options.quantityInWei;
        }
        const { toUnits } = await Promise.resolve().then(() => require("../../../../utils/units.js"));
        return toUnits(options.quantity, await (0, decimals_js_1.decimals)({ contract: options.contract }));
    })();
    const [conditionId, { quantity, currency, pricePerToken, allowlistProof }] = await Promise.all([
        (0, getActiveClaimConditionId_js_1.getActiveClaimConditionId)({
            contract: options.contract,
        }),
        (0, get_claim_params_js_1.getClaimParams)({
            contract: options.contract,
            quantity: quantityWei,
            to: options.claimer,
            type: "erc20",
            from: options.from,
            tokenDecimals: await (0, decimals_js_1.decimals)({ contract: options.contract }),
        }),
    ]);
    try {
        await (0, verifyClaim_js_1.verifyClaim)({
            contract: options.contract,
            claimer: options.claimer,
            quantity,
            currency,
            pricePerToken,
            allowlistProof,
            conditionId,
        });
        return {
            result: true,
        };
    }
    catch (error) {
        return {
            result: false,
            reason: await (0, extract_error_js_1.extractErrorResult)({ error, contract: options.contract }),
        };
    }
}
