// thirdweb/dist/cjs/extensions/erc7579/__generated__/ModularAccountFactory/read/entrypoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isEntrypointSupported = isEntrypointSupported;
exports.decodeEntrypointResult = decodeEntrypointResult;
exports.entrypoint = entrypoint;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa65d69d4";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `entrypoint` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `entrypoint` method is supported.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isEntrypointSupported } from "thirdweb/extensions/erc7579";
 * const supported = isEntrypointSupported(["0x..."]);
 * ```
 */
function isEntrypointSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the entrypoint function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC7579
 * @example
 * ```ts
 * import { decodeEntrypointResult } from "thirdweb/extensions/erc7579";
 * const result = decodeEntrypointResultResult("...");
 * ```
 */
function decodeEntrypointResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "entrypoint" function on the contract.
 * @param options - The options for the entrypoint function.
 * @returns The parsed result of the function call.
 * @extension ERC7579
 * @example
 * ```ts
 * import { entrypoint } from "thirdweb/extensions/erc7579";
 *
 * const result = await entrypoint({
 *  contract,
 * });
 *
 * ```
 */
async function entrypoint(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
