// thirdweb/dist/cjs/react/web/ui/ConnectWallet/Modal/SmartWalletConnectUI.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartConnectUI = SmartConnectUI;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const connection_manager_js_1 = require("../../../../core/providers/connection-manager.js");
const wallet_js_1 = require("../../../../core/utils/wallet.js");
const LoadingScreen_js_1 = require("../../../wallets/shared/LoadingScreen.js");
const getSmartWalletLocale_js_1 = require("../../../wallets/smartWallet/locale/getSmartWalletLocale.js");
const Spacer_js_1 = require("../../components/Spacer.js");
const Spinner_js_1 = require("../../components/Spinner.js");
const basic_js_1 = require("../../components/basic.js");
const text_js_1 = require("../../components/text.js");
const AnyWalletConnectUI_js_1 = require("./AnyWalletConnectUI.js");
/**
 * @internal
 */
function SmartConnectUI(props) {
    const personalWalletInfo = (0, wallet_js_1.useWalletInfo)(props.personalWallet.id);
    const [keyConnected, setKeyConnected] = (0, react_1.useState)(false);
    if (!personalWalletInfo.data) {
        return (0, jsx_runtime_1.jsx)(LoadingScreen_js_1.LoadingScreen, {});
    }
    // connect personal wallet
    if (!keyConnected) {
        return ((0, jsx_runtime_1.jsx)(AnyWalletConnectUI_js_1.AnyWalletConnectUI, { wallet: props.personalWallet, done: () => {
                setKeyConnected(true);
            }, onBack: props.onBack, setModalVisibility: props.setModalVisibility, chain: props.chain, chains: props.chains, client: props.client, meta: props.meta, size: props.size, walletConnect: props.walletConnect, connectLocale: props.connectLocale }));
    }
    return ((0, jsx_runtime_1.jsx)(SmartWalletConnecting, { done: props.done, personalWallet: props.personalWallet, accountAbstraction: props.accountAbstraction, onBack: props.onBack, personalWalletInfo: personalWalletInfo.data, localeId: props.connectLocale.id, size: props.size, client: props.client }));
}
function SmartWalletConnecting(props) {
    const localeQuery = (0, react_query_1.useQuery)({
        queryKey: ["getSmartWalletLocale", props.localeId],
        queryFn: () => (0, getSmartWalletLocale_js_1.getSmartWalletLocale)(props.localeId),
    });
    const { personalWallet } = props;
    const { done } = props;
    const [smartWalletConnectionStatus, setSmartWalletConnectionStatus] = (0, react_1.useState)("idle");
    const connectionManager = (0, connection_manager_js_1.useConnectionManager)();
    const handleConnect = (0, react_1.useCallback)(async () => {
        if (!personalWallet) {
            throw new Error("No personal wallet");
        }
        setSmartWalletConnectionStatus("connecting");
        try {
            const connected = await connectionManager.handleConnection(personalWallet, {
                accountAbstraction: props.accountAbstraction,
                client: props.client,
            });
            done(connected);
            setSmartWalletConnectionStatus("idle");
        }
        catch (e) {
            console.error(e);
            setSmartWalletConnectionStatus("connect-error");
        }
    }, [
        done,
        personalWallet,
        props.client,
        props.accountAbstraction,
        connectionManager,
    ]);
    const connectStarted = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (!connectStarted.current) {
            handleConnect();
            connectStarted.current = true;
        }
    }, [handleConnect]);
    if (!localeQuery.data) {
        return (0, jsx_runtime_1.jsx)(LoadingScreen_js_1.LoadingScreen, {});
    }
    if (smartWalletConnectionStatus === "connect-error") {
        return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { fullHeight: true, animate: "fadein", flex: "column", center: "both", p: "lg", style: {
                minHeight: "300px",
            }, children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "danger", children: localeQuery.data.failedToConnect }) }));
    }
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { fullHeight: true, flex: "column", center: "both", style: {
            minHeight: "300px",
        }, children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", multiline: true, center: true, children: localeQuery.data.connecting }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { color: "accentText", size: "lg" })] }));
}
