// thirdweb/dist/cjs/transaction/prepare-transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionTypeMap = void 0;
exports.prepareTransaction = prepareTransaction;
exports.TransactionTypeMap = {
    legacy: 0,
    eip1559: 1,
    eip2930: 2,
    eip4844: 3,
    eip7702: 4,
};
/**
 * Prepares a transaction with the given options.
 * @param options - The options for preparing the transaction.
 * @param info - Additional information about the ABI function.
 * @returns The prepared transaction.
 * @transaction
 * @example
 * ```ts
 * import { prepareTransaction, toWei } from "thirdweb";
 * import { ethereum } from "thirdweb/chains";
 * const transaction = prepareTransaction({
 *  to: "0x1234567890123456789012345678901234567890",
 *  chain: ethereum,
 *  client: thirdwebClient,
 *  value: toWei("1.0"),
 *  gasPrice: 30n
 * });
 * ```
 */
function prepareTransaction(options, info) {
    if (info) {
        // biome-ignore lint/suspicious/noExplicitAny: TODO: fix later
        options.__preparedMethod = info.preparedMethod;
        // biome-ignore lint/suspicious/noExplicitAny: TODO: fix later
        options.__contract = info.contract;
    }
    return options;
}
