// thirdweb/dist/cjs/extensions/erc4337/account/addSessionKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addSessionKey = addSessionKey;
exports.isAddSessionKeySupported = isAddSessionKeySupported;
exports.shouldUpdateSessionKey = shouldUpdateSessionKey;
const addresses_js_1 = require("../../../constants/addresses.js");
const is_contract_deployed_js_1 = require("../../../utils/bytecode/is-contract-deployed.js");
const units_js_1 = require("../../../utils/units.js");
const getPermissionsForSigner_js_1 = require("../__generated__/IAccountPermissions/read/getPermissionsForSigner.js");
const setPermissionsForSigner_js_1 = require("../__generated__/IAccountPermissions/write/setPermissionsForSigner.js");
const common_js_1 = require("./common.js");
/**
 * Adds session key permissions for a specified address.
 * @param options - The options for the removeSessionKey function.
 * @param {Contract} options.contract - The smart account contract to add the session key to
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { addSessionKey } from 'thirdweb/extensions/erc4337';
 * import { sendTransaction } from 'thirdweb';
 *
 * const transaction = addSessionKey({
 * contract,
 * account,
 * sessionKeyAddress,
 * permissions: {
 *  approvedTargets: ['0x...'],
 *  nativeTokenLimitPerTransaction: 0.1, // in ETH
 *  permissionStartTimestamp: new Date(),
 *  permissionEndTimestamp: new Date(Date.now() + 1000 * 60 * 60 * 24 * 365), // 1 year from now
 * }
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
function addSessionKey(options) {
    const { contract, sessionKeyAddress, account, permissions } = options;
    return (0, setPermissionsForSigner_js_1.setPermissionsForSigner)({
        contract,
        async asyncParams() {
            const { req, signature } = await (0, common_js_1.signPermissionRequest)({
                account,
                contract,
                req: await (0, common_js_1.toContractPermissions)({
                    target: sessionKeyAddress,
                    permissions,
                }),
            });
            return { signature, req };
        },
    });
}
/**
 * Checks if the `isAddSessionKeySupported` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isAddSessionKeySupported` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isAddSessionKeySupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = isAddSessionKeySupported(["0x..."]);
 * ```
 */
function isAddSessionKeySupported(availableSelectors) {
    return (0, setPermissionsForSigner_js_1.isSetPermissionsForSignerSupported)(availableSelectors);
}
/**
 * Checks if the session key should be updated.
 * @param currentPermissions - The current permissions of the session key.
 * @param newPermissions - The new permissions to set for the session key.
 * @returns A boolean indicating if the session key should be updated.
 * @extension ERC4337
 * @example
 * ```ts
 * import { shouldUpdateSessionKey } from "thirdweb/extensions/erc4337";
 *
 * const shouldUpdate = await shouldUpdateSessionKey({ accountContract, sessionKeyAddress, newPermissions });
 * ```
 */
async function shouldUpdateSessionKey(args) {
    const { accountContract, sessionKeyAddress, newPermissions } = args;
    // check if account is deployed
    const accountDeployed = await (0, is_contract_deployed_js_1.isContractDeployed)(accountContract);
    if (!accountDeployed) {
        return true;
    }
    // get current permissions
    const currentPermissions = await (0, getPermissionsForSigner_js_1.getPermissionsForSigner)({
        contract: accountContract,
        signer: sessionKeyAddress,
    });
    // check end time validity
    if (currentPermissions.endTimestamp &&
        currentPermissions.endTimestamp < Math.floor(new Date().getTime() / 1000)) {
        return true;
    }
    // check targets
    if (!areSessionKeyContractTargetsEqual(currentPermissions.approvedTargets, newPermissions.approvedTargets)) {
        return true;
    }
    // check if the new native token limit is greater than the current one
    if ((0, units_js_1.toWei)(newPermissions.nativeTokenLimitPerTransaction?.toString() ?? "0") >
        currentPermissions.nativeTokenLimitPerTransaction) {
        return true;
    }
    return false;
}
function areSessionKeyContractTargetsEqual(currentTargets, newTargets) {
    // Handle the case where approvedTargets is "*"
    if (newTargets === "*" &&
        currentTargets.length === 1 &&
        currentTargets[0] === addresses_js_1.ZERO_ADDRESS) {
        return true;
    }
    if (newTargets !== "*") {
        return newTargets
            .map((target) => target.toLowerCase())
            .every((target) => currentTargets.map((t) => t.toLowerCase()).includes(target));
    }
    return false;
}
