// thirdweb/dist/cjs/utils/url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHttpUrl = isHttpUrl;
exports.formatUniversalUrl = formatUniversalUrl;
exports.formatNativeUrl = formatNativeUrl;
exports.formatWalletConnectUrl = formatWalletConnectUrl;
exports.formatExplorerTxUrl = formatExplorerTxUrl;
exports.formatExplorerAddressUrl = formatExplorerAddressUrl;
/**
 * @internal
 */
function isHttpUrl(url) {
    return url.startsWith("http://") || url.startsWith("https://");
}
/**
 * @internal
 */
function formatUniversalUrl(appUrl, wcUri) {
    if (!isHttpUrl(appUrl)) {
        return formatNativeUrl(appUrl, wcUri);
    }
    let safeAppUrl = appUrl;
    if (!safeAppUrl.endsWith("/")) {
        safeAppUrl = `${safeAppUrl}/`;
    }
    const encodedWcUrl = encodeURIComponent(wcUri);
    return {
        redirect: `${safeAppUrl}wc?uri=${encodedWcUrl}`,
        href: safeAppUrl,
    };
}
/**
 * @internal
 */
function formatNativeUrl(appUrl, wcUri) {
    if (isHttpUrl(appUrl)) {
        return formatUniversalUrl(appUrl, wcUri);
    }
    let safeAppUrl = appUrl;
    if (!safeAppUrl.includes("://")) {
        safeAppUrl = appUrl.replaceAll("/", "").replaceAll(":", "");
        safeAppUrl = `${safeAppUrl}://`;
    }
    if (!safeAppUrl.endsWith("/")) {
        safeAppUrl = `${safeAppUrl}/`;
    }
    const encodedWcUrl = encodeURIComponent(wcUri);
    return {
        redirect: `${safeAppUrl}wc?uri=${encodedWcUrl}`,
        href: safeAppUrl,
    };
}
/**
 * @internal
 */
function formatWalletConnectUrl(appUrl, wcUri) {
    return isHttpUrl(appUrl)
        ? formatUniversalUrl(appUrl, wcUri)
        : formatNativeUrl(appUrl, wcUri);
}
function formatExplorerTxUrl(explorerUrl, txHash) {
    return `${explorerUrl.endsWith("/") ? explorerUrl : `${explorerUrl}/`}tx/${txHash}`;
}
function formatExplorerAddressUrl(explorerUrl, address) {
    return `${explorerUrl.endsWith("/") ? explorerUrl : `${explorerUrl}/`}address/${address}`;
}
