// thirdweb/dist/esm/extensions/erc7579/__generated__/ModularAccountFactory/read/entrypoint.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xa65d69d4";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `entrypoint` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `entrypoint` method is supported.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isEntrypointSupported } from "thirdweb/extensions/erc7579";
 * const supported = isEntrypointSupported(["0x..."]);
 * ```
 */
export function isEntrypointSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the entrypoint function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC7579
 * @example
 * ```ts
 * import { decodeEntrypointResult } from "thirdweb/extensions/erc7579";
 * const result = decodeEntrypointResultResult("...");
 * ```
 */
export function decodeEntrypointResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "entrypoint" function on the contract.
 * @param options - The options for the entrypoint function.
 * @returns The parsed result of the function call.
 * @extension ERC7579
 * @example
 * ```ts
 * import { entrypoint } from "thirdweb/extensions/erc7579";
 *
 * const result = await entrypoint({
 *  contract,
 * });
 *
 * ```
 */
export async function entrypoint(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
