// thirdweb/dist/esm/extensions/erc7579/__generated__/ModularAccountFactory/read/implementation.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x5c60da1b";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
        name: "result",
    },
];
/**
 * Checks if the `implementation` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `implementation` method is supported.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isImplementationSupported } from "thirdweb/extensions/erc7579";
 * const supported = isImplementationSupported(["0x..."]);
 * ```
 */
export function isImplementationSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the implementation function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC7579
 * @example
 * ```ts
 * import { decodeImplementationResult } from "thirdweb/extensions/erc7579";
 * const result = decodeImplementationResultResult("...");
 * ```
 */
export function decodeImplementationResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "implementation" function on the contract.
 * @param options - The options for the implementation function.
 * @returns The parsed result of the function call.
 * @extension ERC7579
 * @example
 * ```ts
 * import { implementation } from "thirdweb/extensions/erc7579";
 *
 * const result = await implementation({
 *  contract,
 * });
 *
 * ```
 */
export async function implementation(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
