// thirdweb/dist/cjs/analytics/track/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackTransaction = trackTransaction;
const json_js_1 = require("../../utils/json.js");
const index_js_1 = require("./index.js");
/**
 * @internal
 */
async function trackTransaction(args) {
    return trackTransactionEvent({
        ...args,
        action: "transaction:sent",
    });
}
/**
 * @internal
 */
function trackTransactionEvent(args) {
    return (0, index_js_1.track)({
        client: args.client,
        ecosystem: args.ecosystem,
        data: {
            action: args.action,
            clientId: args.client.clientId,
            chainId: args.chainId,
            transactionHash: args.transactionHash,
            walletAddress: args.walletAddress,
            walletType: args.walletType,
            contractAddress: args.contractAddress,
            functionName: args.functionName,
            gasPrice: args.gasPrice,
            errorCode: (0, json_js_1.stringify)(args.error),
        },
    });
}
