// thirdweb/dist/cjs/wallets/smart/smart-wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.smartWallet = smartWallet;
const connect_js_1 = require("../../analytics/track/connect.js");
const utils_js_1 = require("../../chains/utils.js");
const contract_js_1 = require("../../contract/contract.js");
const isZkSyncChain_js_1 = require("../../utils/any-evm/zksync/isZkSyncChain.js");
const is_contract_deployed_js_1 = require("../../utils/bytecode/is-contract-deployed.js");
const wallet_emitter_js_1 = require("../wallet-emitter.js");
const constants_js_1 = require("./lib/constants.js");
/**
 * Creates a ERC4337 smart wallet based on a admin account.
 *
 * Smart wallets are smart contract wallets that enable multiple benefits for users:
 *
 * - Sponsor gas fees for transactions
 * - Multiple owners
 * - Session keys
 * - Batch transactions
 * - Predictable addresses
 * - Programmable features
 *
 * [Learn more about account abstraction](https://portal.thirdweb.com/connect/account-abstraction/how-it-works)
 *
 * @param createOptions - The options for creating the wallet.
 * Refer to [SmartWalletCreationOptions](https://portal.thirdweb.com/references/typescript/v5/SmartWalletCreationOptions) for more details.
 * @returns The created smart wallet.
 * @example
 *
 * ## Connect to a smart wallet
 *
 * To connect to a smart wallet, you need to provide an admin account as the `personalAccount` option.
 *
 * Any wallet can be used as an admin account, including an in-app wallets.
 *
 * The `sponsorGas` option is used to enable sponsored gas for transactions automatically.
 *
 * ```ts
 * import { smartWallet, inAppWallet } from "thirdweb/wallets";
 * import { sepolia } from "thirdweb/chains";
 * import { sendTransaction } from "thirdweb";
 *
 * const wallet = smartWallet({
 *  chain: sepolia,
 *  sponsorGas: true, // enable sponsored transactions
 * });
 *
 * // any wallet can be used as an admin account
 * // in this example we use an in-app wallet
 * const adminWallet = inAppWallet();
 * const personalAccount = await adminWallet.connect({
 *   client,
 *   chain: sepolia,
 *   strategy: "google",
 * });
 *
 * const smartAccount = await wallet.connect({
 *   client,
 *   personalAccount, // pass the admin account
 * });
 *
 * // sending sponsored transactions with the smartAccount
 * await sendTransaction({
 *   account: smartAccount,
 *   transaction,
 * });
 * ```
 *
 * ## Using a custom account factory
 *
 * You can pass a custom account factory to the `smartWallet` function to use a your own account factory.
 *
 * ```ts
 * import { smartWallet } from "thirdweb/wallets";
 * import { sepolia } from "thirdweb/chains";
 *
 * const wallet = smartWallet({
 *  chain: sepolia,
 *  sponsorGas: true, // enable sponsored transactions
 *  factoryAddress: "0x...", // custom factory address
 * });
 * ```
 *
 * ## Using v0.7 Entrypoint
 *
 * Both v0.6 (default) and v0.7 ERC4337 Entrypoints are supported. To use the v0.7 Entrypoint, simply pass in a compatible account factory.
 *
 * You can use the predeployed `DEFAULT_ACCOUNT_FACTORY_V0_7` or deploy your own [AccountFactory  v0.7](https://thirdweb.com/thirdweb.eth/AccountFactory_0_7).
 *
 * ```ts
 * import { smartWallet, DEFAULT_ACCOUNT_FACTORY_V0_7 } from "thirdweb/wallets/smart";
 * import { sepolia } from "thirdweb/chains";
 *
 * const wallet = smartWallet({
 *  chain: sepolia,
 *  sponsorGas: true, // enable sponsored transactions
 *  factoryAddress: DEFAULT_ACCOUNT_FACTORY_V0_7, // 0.7 factory address
 * });
 * ```
 *
 * ## Configuring the smart wallet
 *
 * You can pass options to the `smartWallet` function to configure the smart wallet.
 *
 * ```ts
 * import { smartWallet } from "thirdweb/wallets";
 * import { sepolia } from "thirdweb/chains";
 *
 * const wallet = smartWallet({
 *  chain: sepolia,
 *  sponsorGas: true, // enable sponsored transactions
 *  factoryAddress: "0x...", // custom factory address
 *  overrides: {
 *    accountAddress: "0x...", // override account address
 *    accountSalt: "0x...", // override account salt
 *    entrypointAddress: "0x...", // override entrypoint address
 *    tokenPaymaster: TokenPaymaster.BASE_USDC, // enable erc20 paymaster
 *    bundlerUrl: "https://...", // override bundler url
 *    paymaster: (userOp) => { ... }, // override paymaster
 *    ...
 *  }
 * });
 * ```
 *
 * Refer to [SmartWalletOptions](https://portal.thirdweb.com/references/typescript/v5/SmartWalletOptions) for more details.
 *
 * @wallet
 */
function smartWallet(createOptions) {
    const emitter = (0, wallet_emitter_js_1.createWalletEmitter)();
    let account = undefined;
    let adminAccount = undefined;
    let chain = undefined;
    let lastConnectOptions;
    return {
        id: "smart",
        subscribe: emitter.subscribe,
        getChain() {
            if (!chain) {
                return undefined;
            }
            chain = (0, utils_js_1.getCachedChainIfExists)(chain.id) || chain;
            return chain;
        },
        getConfig: () => createOptions,
        getAccount: () => account,
        getAdminAccount: () => adminAccount,
        autoConnect: async (options) => {
            const { connectSmartAccount: connectSmartWallet } = await Promise.resolve().then(() => require("./index.js"));
            const [connectedAccount, connectedChain] = await connectSmartWallet(options, createOptions);
            // set the states
            lastConnectOptions = options;
            account = connectedAccount;
            chain = connectedChain;
            (0, connect_js_1.trackConnect)({
                client: options.client,
                walletType: "smart",
                walletAddress: account.address,
                chainId: chain.id,
            });
            // return account
            return account;
        },
        connect: async (options) => {
            const { connectSmartAccount } = await Promise.resolve().then(() => require("./index.js"));
            const [connectedAccount, connectedChain] = await connectSmartAccount(options, createOptions);
            // set the states
            adminAccount = options.personalAccount;
            lastConnectOptions = options;
            account = connectedAccount;
            chain = connectedChain;
            (0, connect_js_1.trackConnect)({
                client: options.client,
                walletType: "smart",
                walletAddress: account.address,
                chainId: chain.id,
            });
            // return account
            emitter.emit("accountChanged", account);
            return account;
        },
        disconnect: async () => {
            if (account) {
                const { disconnectSmartAccount } = await Promise.resolve().then(() => require("./index.js"));
                await disconnectSmartAccount(account);
            }
            account = undefined;
            adminAccount = undefined;
            chain = undefined;
            emitter.emit("disconnect", undefined);
        },
        switchChain: async (newChain) => {
            if (!lastConnectOptions) {
                throw new Error("Cannot switch chain without a previous connection");
            }
            const isZksyncChain = await (0, isZkSyncChain_js_1.isZkSyncChain)(newChain);
            if (!isZksyncChain) {
                // check if factory is deployed
                const factory = (0, contract_js_1.getContract)({
                    address: createOptions.factoryAddress ||
                        (0, constants_js_1.getDefaultAccountFactory)(createOptions.overrides?.entrypointAddress),
                    chain: newChain,
                    client: lastConnectOptions.client,
                });
                const isDeployed = await (0, is_contract_deployed_js_1.isContractDeployed)(factory);
                if (!isDeployed) {
                    throw new Error(`Factory contract not deployed on chain: ${newChain.id}`);
                }
            }
            const { connectSmartAccount } = await Promise.resolve().then(() => require("./index.js"));
            const [connectedAccount, connectedChain] = await connectSmartAccount({ ...lastConnectOptions, chain: newChain }, createOptions);
            // set the states
            account = connectedAccount;
            chain = connectedChain;
            emitter.emit("accountChanged", connectedAccount);
            emitter.emit("chainChanged", connectedChain);
        },
    };
}
