// thirdweb/dist/esm/extensions/erc7579/__generated__/IERC7579Account/read/isModuleInstalled.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x112d3a7d";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "moduleTypeId",
    },
    {
        type: "address",
        name: "module",
    },
    {
        type: "bytes",
        name: "additionalContext",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isModuleInstalled` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isModuleInstalled` method is supported.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isIsModuleInstalledSupported } from "thirdweb/extensions/erc7579";
 * const supported = isIsModuleInstalledSupported(["0x..."]);
 * ```
 */
export function isIsModuleInstalledSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isModuleInstalled" function.
 * @param options - The options for the isModuleInstalled function.
 * @returns The encoded ABI parameters.
 * @extension ERC7579
 * @example
 * ```ts
 * import { encodeIsModuleInstalledParams } from "thirdweb/extensions/erc7579";
 * const result = encodeIsModuleInstalledParams({
 *  moduleTypeId: ...,
 *  module: ...,
 *  additionalContext: ...,
 * });
 * ```
 */
export function encodeIsModuleInstalledParams(options) {
    return encodeAbiParameters(FN_INPUTS, [
        options.moduleTypeId,
        options.module,
        options.additionalContext,
    ]);
}
/**
 * Encodes the "isModuleInstalled" function into a Hex string with its parameters.
 * @param options - The options for the isModuleInstalled function.
 * @returns The encoded hexadecimal string.
 * @extension ERC7579
 * @example
 * ```ts
 * import { encodeIsModuleInstalled } from "thirdweb/extensions/erc7579";
 * const result = encodeIsModuleInstalled({
 *  moduleTypeId: ...,
 *  module: ...,
 *  additionalContext: ...,
 * });
 * ```
 */
export function encodeIsModuleInstalled(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeIsModuleInstalledParams(options).slice(2));
}
/**
 * Decodes the result of the isModuleInstalled function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC7579
 * @example
 * ```ts
 * import { decodeIsModuleInstalledResult } from "thirdweb/extensions/erc7579";
 * const result = decodeIsModuleInstalledResultResult("...");
 * ```
 */
export function decodeIsModuleInstalledResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isModuleInstalled" function on the contract.
 * @param options - The options for the isModuleInstalled function.
 * @returns The parsed result of the function call.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isModuleInstalled } from "thirdweb/extensions/erc7579";
 *
 * const result = await isModuleInstalled({
 *  contract,
 *  moduleTypeId: ...,
 *  module: ...,
 *  additionalContext: ...,
 * });
 *
 * ```
 */
export async function isModuleInstalled(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.moduleTypeId, options.module, options.additionalContext],
    });
}
