// thirdweb/dist/cjs/extensions/modules/__generated__/SequentialTokenIdERC1155/read/getModuleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetModuleConfigSupported = isGetModuleConfigSupported;
exports.decodeGetModuleConfigResult = decodeGetModuleConfigResult;
exports.getModuleConfig = getModuleConfig;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x89e04e0e";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple",
        name: "config",
        components: [
            {
                type: "bool",
                name: "registerInstallationCallback",
            },
            {
                type: "bytes4[]",
                name: "requiredInterfaces",
            },
            {
                type: "bytes4[]",
                name: "supportedInterfaces",
            },
            {
                type: "tuple[]",
                name: "callbackFunctions",
                components: [
                    {
                        type: "bytes4",
                        name: "selector",
                    },
                ],
            },
            {
                type: "tuple[]",
                name: "fallbackFunctions",
                components: [
                    {
                        type: "bytes4",
                        name: "selector",
                    },
                    {
                        type: "uint256",
                        name: "permissionBits",
                    },
                ],
            },
        ],
    },
];
/**
 * Checks if the `getModuleConfig` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getModuleConfig` method is supported.
 * @modules SequentialTokenIdERC1155
 * @example
 * ```ts
 * import { SequentialTokenIdERC1155 } from "thirdweb/modules";
 * const supported = SequentialTokenIdERC1155.isGetModuleConfigSupported(["0x..."]);
 * ```
 */
function isGetModuleConfigSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getModuleConfig function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules SequentialTokenIdERC1155
 * @example
 * ```ts
 * import { SequentialTokenIdERC1155 } from "thirdweb/modules";
 * const result = SequentialTokenIdERC1155.decodeGetModuleConfigResultResult("...");
 * ```
 */
function decodeGetModuleConfigResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getModuleConfig" function on the contract.
 * @param options - The options for the getModuleConfig function.
 * @returns The parsed result of the function call.
 * @modules SequentialTokenIdERC1155
 * @example
 * ```ts
 * import { SequentialTokenIdERC1155 } from "thirdweb/modules";
 *
 * const result = await SequentialTokenIdERC1155.getModuleConfig({
 *  contract,
 * });
 *
 * ```
 */
async function getModuleConfig(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
