// thirdweb/dist/cjs/extensions/stylus/write/activateStylusContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activateStylusContract = activateStylusContract;
const viem_1 = require("viem");
const contract_js_1 = require("../../../contract/contract.js");
const eth_call_js_1 = require("../../../rpc/actions/eth_call.js");
const rpc_js_1 = require("../../../rpc/rpc.js");
const encode_js_1 = require("../../../transaction/actions/encode.js");
const activateProgram_js_1 = require("../__generated__/IArbWasm/write/activateProgram.js");
const ARB_WASM_ADDRESS = "0x0000000000000000000000000000000000000071";
/**
 * Activate a stylus contract by calling ArbWasm Precompile
 * @param options - The options for activating the contract
 * @returns A prepared transaction to send
 * @example
 * ```ts
 * import { activateStylusContract } from "thirdweb/stylus";
 * const transaction = activateStylusContract({
 *  client,
 *  chain,
 *  contractAddress,
 * });
 * await sendTransaction({ transaction, account });
 * ```
 */
async function activateStylusContract(options) {
    const { chain, client, contractAddress } = options;
    const arbWasmPrecompile = (0, contract_js_1.getContract)({
        client,
        chain,
        address: ARB_WASM_ADDRESS,
    });
    const dataFee = await estimateDataFee({
        transaction: (0, activateProgram_js_1.activateProgram)({
            program: contractAddress,
            contract: arbWasmPrecompile,
        }),
    });
    return (0, activateProgram_js_1.activateProgram)({
        program: contractAddress,
        contract: arbWasmPrecompile,
        overrides: {
            value: dataFee,
        },
    });
}
async function estimateDataFee(options) {
    const data = await (0, encode_js_1.encode)(options.transaction);
    const serializedTx = (0, viem_1.formatTransactionRequest)({
        data,
        to: ARB_WASM_ADDRESS,
        value: (0, viem_1.parseEther)("1"), // only for simulation. it will be replaced with estimated data fee.
    });
    const rpcRequest = (0, rpc_js_1.getRpcClient)(options.transaction);
    try {
        const result = await (0, eth_call_js_1.eth_call)(rpcRequest, serializedTx);
        const [, dataFee] = (0, viem_1.decodeAbiParameters)([
            {
                type: "uint16",
            },
            {
                type: "uint256",
            },
        ], result);
        return (dataFee * BigInt(100 + 10)) / BigInt(100); // bump 10%
    }
    catch {
        return 0n;
    }
}
