// thirdweb/dist/cjs/cli/commands/stylus/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStylusProject = createStylusProject;
const node_child_process_1 = require("node:child_process");
const ora_1 = require("ora");
const prompts_1 = require("prompts");
async function createStylusProject() {
    const spinner = (0, ora_1.default)();
    // Step 1: Ensure cargo is installed
    const cargoCheck = (0, node_child_process_1.spawnSync)("cargo", ["--version"]);
    if (cargoCheck.status !== 0) {
        console.error("Error: `cargo` is not installed");
        process.exit(1);
    }
    // Step 2: Install stylus etc.
    spinner.start("Installing Stylus...");
    const install = (0, node_child_process_1.spawnSync)("cargo", ["install", "--force", "cargo-stylus"], {
        stdio: "inherit",
    });
    if (install.status !== 0) {
        spinner.fail("Failed to install Stylus.");
        process.exit(1);
    }
    spinner.succeed("Stylus installed.");
    (0, node_child_process_1.spawnSync)("rustup", ["default", "1.83"], {
        stdio: "inherit",
    });
    (0, node_child_process_1.spawnSync)("rustup", ["target", "add", "wasm32-unknown-unknown"], {
        stdio: "inherit",
    });
    // Step 3: Create the project
    const { projectName } = await (0, prompts_1.default)({
        type: "text",
        name: "projectName",
        message: "Project name:",
        initial: "my-stylus-project",
    });
    spinner.start(`Creating new Stylus project: ${projectName}...`);
    const newProject = (0, node_child_process_1.spawnSync)("cargo", ["stylus", "new", projectName], {
        stdio: "inherit",
    });
    if (newProject.status !== 0) {
        spinner.fail("Failed to create Stylus project.");
        process.exit(1);
    }
    spinner.succeed("Project created successfully.");
    console.log(`\n✅ Done! cd into your project: ${projectName}`);
}
