// thirdweb/dist/cjs/extensions/erc7579/__generated__/IERC7579Account/read/supportsModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isSupportsModuleSupported = isSupportsModuleSupported;
exports.encodeSupportsModuleParams = encodeSupportsModuleParams;
exports.encodeSupportsModule = encodeSupportsModule;
exports.decodeSupportsModuleResult = decodeSupportsModuleResult;
exports.supportsModule = supportsModule;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xf2dc691d";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "moduleTypeId",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `supportsModule` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `supportsModule` method is supported.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isSupportsModuleSupported } from "thirdweb/extensions/erc7579";
 * const supported = isSupportsModuleSupported(["0x..."]);
 * ```
 */
function isSupportsModuleSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "supportsModule" function.
 * @param options - The options for the supportsModule function.
 * @returns The encoded ABI parameters.
 * @extension ERC7579
 * @example
 * ```ts
 * import { encodeSupportsModuleParams } from "thirdweb/extensions/erc7579";
 * const result = encodeSupportsModuleParams({
 *  moduleTypeId: ...,
 * });
 * ```
 */
function encodeSupportsModuleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.moduleTypeId]);
}
/**
 * Encodes the "supportsModule" function into a Hex string with its parameters.
 * @param options - The options for the supportsModule function.
 * @returns The encoded hexadecimal string.
 * @extension ERC7579
 * @example
 * ```ts
 * import { encodeSupportsModule } from "thirdweb/extensions/erc7579";
 * const result = encodeSupportsModule({
 *  moduleTypeId: ...,
 * });
 * ```
 */
function encodeSupportsModule(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeSupportsModuleParams(options).slice(2));
}
/**
 * Decodes the result of the supportsModule function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC7579
 * @example
 * ```ts
 * import { decodeSupportsModuleResult } from "thirdweb/extensions/erc7579";
 * const result = decodeSupportsModuleResultResult("...");
 * ```
 */
function decodeSupportsModuleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "supportsModule" function on the contract.
 * @param options - The options for the supportsModule function.
 * @returns The parsed result of the function call.
 * @extension ERC7579
 * @example
 * ```ts
 * import { supportsModule } from "thirdweb/extensions/erc7579";
 *
 * const result = await supportsModule({
 *  contract,
 *  moduleTypeId: ...,
 * });
 *
 * ```
 */
async function supportsModule(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.moduleTypeId],
    });
}
