// thirdweb/dist/cjs/extensions/erc7579/__generated__/ModularAccountFactory/write/upgradeTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isUpgradeToSupported = isUpgradeToSupported;
exports.encodeUpgradeToParams = encodeUpgradeToParams;
exports.encodeUpgradeTo = encodeUpgradeTo;
exports.upgradeTo = upgradeTo;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3659cfe6";
const FN_INPUTS = [
    {
        type: "address",
        name: "newImplementation",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `upgradeTo` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `upgradeTo` method is supported.
 * @extension ERC7579
 * @example
 * ```ts
 * import { isUpgradeToSupported } from "thirdweb/extensions/erc7579";
 *
 * const supported = isUpgradeToSupported(["0x..."]);
 * ```
 */
function isUpgradeToSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "upgradeTo" function.
 * @param options - The options for the upgradeTo function.
 * @returns The encoded ABI parameters.
 * @extension ERC7579
 * @example
 * ```ts
 * import { encodeUpgradeToParams } from "thirdweb/extensions/erc7579";
 * const result = encodeUpgradeToParams({
 *  newImplementation: ...,
 * });
 * ```
 */
function encodeUpgradeToParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.newImplementation]);
}
/**
 * Encodes the "upgradeTo" function into a Hex string with its parameters.
 * @param options - The options for the upgradeTo function.
 * @returns The encoded hexadecimal string.
 * @extension ERC7579
 * @example
 * ```ts
 * import { encodeUpgradeTo } from "thirdweb/extensions/erc7579";
 * const result = encodeUpgradeTo({
 *  newImplementation: ...,
 * });
 * ```
 */
function encodeUpgradeTo(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeUpgradeToParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "upgradeTo" function on the contract.
 * @param options - The options for the "upgradeTo" function.
 * @returns A prepared transaction object.
 * @extension ERC7579
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { upgradeTo } from "thirdweb/extensions/erc7579";
 *
 * const transaction = upgradeTo({
 *  contract,
 *  newImplementation: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
function upgradeTo(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.newImplementation];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
    });
}
