// thirdweb/dist/cjs/react/native/ui/prebuilt/Account/blobbie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountBlobbie = AccountBlobbie;
exports.Blobbie = Blobbie;
const jsx_runtime_1 = require("react/jsx-runtime");
const utils_1 = require("@noble/curves/abstract/utils");
const react_1 = require("react");
const react_native_1 = require("react-native");
const react_native_svg_1 = require("react-native-svg");
const provider_js_1 = require("../../../../core/account/provider.js");
const account_js_1 = require("../../../../core/utils/account.js");
/**
 * A wrapper for the Blobbie component
 * @param props BlobbieProps
 * @beta
 * @wallet
 */
function AccountBlobbie(props) {
    const { address } = (0, provider_js_1.useAccountContext)();
    return (0, jsx_runtime_1.jsx)(Blobbie, { ...props, address: address });
}
/**
 * A unique gradient avatar based on the provided address.
 * @param props The component props.
 * @param props.address The address to generate the gradient with.
 * @param props.style The style for the component
 * @param props.size The size of each side of the square avatar (in pixels)
 * @component
 * @wallet
 * @example
 * ```tsx
 * import { Blobbie } from "thirdweb/react";
 *
 * <Blobbie address="0x...." style={{ width: 40, height: 40 }} />
 * ```
 */
function Blobbie(props) {
    const colors = (0, react_1.useMemo)(() => account_js_1.COLOR_OPTIONS[Number((0, utils_1.hexToNumber)(props.address.slice(2, 4))) % account_js_1.COLOR_OPTIONS.length], [props.address]);
    const containerStyle = (0, react_1.useMemo)(() => {
        const baseStyle = props.style || {};
        if (props.size) {
            return {
                ...baseStyle,
                width: props.size,
                height: props.size,
            };
        }
        return baseStyle;
    }, [props.style, props.size]);
    const gradientUniqueId = `grad${colors[0]}+${colors[1]}`.replace(/[^a-zA-Z0-9 ]/g, "");
    return ((0, jsx_runtime_1.jsx)(react_native_1.View, { style: [styles.container, containerStyle], children: (0, jsx_runtime_1.jsxs)(react_native_svg_1.Svg, { height: "100%", width: "100%", style: react_native_1.StyleSheet.absoluteFillObject, children: [(0, jsx_runtime_1.jsx)(react_native_svg_1.Defs, { children: (0, jsx_runtime_1.jsxs)(react_native_svg_1.LinearGradient, { id: gradientUniqueId, x1: "0%", y1: "100%", x2: "100%", y2: "0%", children: [(0, jsx_runtime_1.jsx)(react_native_svg_1.Stop, { offset: "0", stopColor: colors[0] }), (0, jsx_runtime_1.jsx)(react_native_svg_1.Stop, { offset: "1", stopColor: colors[1] })] }) }), (0, jsx_runtime_1.jsx)(react_native_svg_1.Rect, { width: "100%", height: "100%", fill: `url(#${gradientUniqueId})` })] }) }));
}
const styles = react_native_1.StyleSheet.create({
    container: {
        overflow: "hidden",
    },
});
