// thirdweb/dist/esm/utils/url.js
/**
 * @internal
 */
export function isHttpUrl(url) {
    return url.startsWith("http://") || url.startsWith("https://");
}
/**
 * @internal
 */
export function formatUniversalUrl(appUrl, wcUri) {
    if (!isHttpUrl(appUrl)) {
        return formatNativeUrl(appUrl, wcUri);
    }
    let safeAppUrl = appUrl;
    if (!safeAppUrl.endsWith("/")) {
        safeAppUrl = `${safeAppUrl}/`;
    }
    const encodedWcUrl = encodeURIComponent(wcUri);
    return {
        redirect: `${safeAppUrl}wc?uri=${encodedWcUrl}`,
        href: safeAppUrl,
    };
}
/**
 * @internal
 */
export function formatNativeUrl(appUrl, wcUri) {
    if (isHttpUrl(appUrl)) {
        return formatUniversalUrl(appUrl, wcUri);
    }
    let safeAppUrl = appUrl;
    if (!safeAppUrl.includes("://")) {
        safeAppUrl = appUrl.replaceAll("/", "").replaceAll(":", "");
        safeAppUrl = `${safeAppUrl}://`;
    }
    if (!safeAppUrl.endsWith("/")) {
        safeAppUrl = `${safeAppUrl}/`;
    }
    const encodedWcUrl = encodeURIComponent(wcUri);
    return {
        redirect: `${safeAppUrl}wc?uri=${encodedWcUrl}`,
        href: safeAppUrl,
    };
}
/**
 * @internal
 */
export function formatWalletConnectUrl(appUrl, wcUri) {
    return isHttpUrl(appUrl)
        ? formatUniversalUrl(appUrl, wcUri)
        : formatNativeUrl(appUrl, wcUri);
}
export function formatExplorerTxUrl(explorerUrl, txHash) {
    return `${explorerUrl.endsWith("/") ? explorerUrl : `${explorerUrl}/`}tx/${txHash}`;
}
export function formatExplorerAddressUrl(explorerUrl, address) {
    return `${explorerUrl.endsWith("/") ? explorerUrl : `${explorerUrl}/`}address/${address}`;
}
