// thirdweb/dist/esm/react/web/ui/prebuilt/Account/balance.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import { useActiveWalletChain } from "../../../../../react/core/hooks/wallets/useActiveWalletChain.js";
import { getAddress } from "../../../../../utils/address.js";
import { useAccountContext } from "../../../../core/account/provider.js";
import { formatAccountFiatBalance, loadAccountBalance, } from "../../../../core/utils/account.js";
import { formatAccountTokenBalance } from "../../../../core/utils/account.js";
/**
 * This component fetches and shows the balance of the wallet address on a given chain.
 * It inherits all the attributes of a HTML <span> component, hence you can style it just like how you would style a normal <span>
 *
 *
 * @example
 * ### Basic usage
 * ```tsx
 * import { AccountProvider, AccountBalance } from "thirdweb/react";
 * import { ethereum } from "thirdweb/chains";
 *
 * <AccountProvider address="0x...">
 *   <AccountBalance chain={ethereum} />
 * </AccountProvider>
 * ```
 * Result:
 * ```html
 * <span>1.091435 ETH</span>
 * ```
 *
 *
 * ### Format the balance (round up, shorten etc.)
 * The AccountBalance component accepts a `formatFn` which takes in an object of type `AccountBalanceInfo` and outputs a string
 * The function is used to modify the display value of the wallet balance (either in crypto or fiat)
 *
 * ```tsx
 * import type { AccountBalanceInfo } from "thirdweb/react";
 * import { formatNumber } from "thirdweb/utils";
 *
 * const format = (props: AccountInfoBalance):string => `${formatNumber(props.balance, 1)} ${props.symbol.toLowerCase()}`
 *
 * <AccountBalance formatFn={format} />
 * ```
 *
 * Result:
 * ```html
 * <span>1.1 eth</span> // the balance is rounded up to 1 decimal and the symbol is lowercased
 * ```
 *
 * ### Show a loading sign when the balance is being fetched
 * ```tsx
 * import { AccountProvider, AccountBalance } from "thirdweb/react";
 *
 * <AccountProvider address="0x...">
 *   <AccountBalance
 *     chain={ethereum}
 *     loadingComponent={<Spinner />}
 *   />
 * </AccountProvider>
 * ```
 *
 * ### Fallback to something when the balance fails to resolve
 * ```tsx
 * <AccountProvider address="0x...">
 *   <AccountBalance
 *     chain={nonExistentChain}
 *     fallbackComponent={"Failed to load"}
 *   />
 * </AccountProvider>
 * ```
 *
 * ### Custom query options for useQuery
 * This component uses `@tanstack-query`'s useQuery internally.
 * You can use the `queryOptions` prop for more fine-grained control
 * ```tsx
 * <AccountBalance
 *   queryOptions={{
 *     enabled: isEnabled,
 *     retry: 4,
 *   }}
 * />
 * ```
 *
 * @component
 * @wallet
 * @beta
 */
export function AccountBalance({ chain, tokenAddress, loadingComponent, fallbackComponent, queryOptions, formatFn, showBalanceInFiat, ...restProps }) {
    const { address, client } = useAccountContext();
    const walletChain = useActiveWalletChain();
    const chainToLoad = chain || walletChain;
    const balanceQuery = useQuery({
        queryKey: [
            "internal_account_balance",
            chainToLoad?.id || -1,
            address,
            { tokenAddress },
            showBalanceInFiat,
        ],
        queryFn: async () => loadAccountBalance({
            chain: chainToLoad,
            client,
            address: getAddress(address),
            tokenAddress: tokenAddress ? getAddress(tokenAddress) : undefined,
            showBalanceInFiat,
        }),
        retry: false,
        ...queryOptions,
    });
    if (balanceQuery.isLoading) {
        return loadingComponent || null;
    }
    if (!balanceQuery.data) {
        return fallbackComponent || null;
    }
    // Prioritize using the formatFn from users
    if (formatFn) {
        return _jsx("span", { ...restProps, children: formatFn(balanceQuery.data) });
    }
    if (showBalanceInFiat) {
        return (_jsx("span", { ...restProps, children: formatAccountFiatBalance({ ...balanceQuery.data, decimals: 0 }) }));
    }
    return (_jsx("span", { ...restProps, children: formatAccountTokenBalance({
            ...balanceQuery.data,
            decimals: balanceQuery.data.balance < 1 ? 3 : 2,
        }) }));
}
